// mkcentral-api/dist/lib/common/GameMode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameMode = void 0;
var Helpers_1 = require("./Helpers");
// Central mode registry (single source of truth)
/**
 * Central registry of all known game modes keyed by {@link GameModeKey}.
 */
var MODE_DEFINITIONS = Object.freeze({
    '150cc': { displayName: '150cc', shortName: '150cc' },
    '200cc': { displayName: '200cc', shortName: '200cc' },
    rt: { displayName: 'Regular Tracks', shortName: 'RT' },
    ct: { displayName: 'Custom Tracks', shortName: 'CT' },
    vsrace: { displayName: 'VS Race', shortName: 'VS' },
    match_race: { displayName: 'Match Race', shortName: 'MR' },
    mixed_battle: { displayName: 'Mixed Battle', shortName: 'Mixed Battle' },
    balloon_battle: { displayName: 'Balloon Battle', shortName: 'Balloon' },
    shine_thief: { displayName: 'Shine Thief', shortName: 'Shine' },
    bobomb_blast: { displayName: 'Bob-omb Blast', shortName: 'Bob-omb' },
    coin_runners: { displayName: 'Coin Runners', shortName: 'Coins' },
    renegade_roundup: { displayName: 'Renegade Roundup', shortName: 'Renegade' },
    mixed: { displayName: 'Mixed', shortName: 'Mixed' },
});
/**
 * Immutable game mode metadata (display/short names).
 * Construct via key or use the provided static getters (e.g., {@link GameMode.RT}).
 */
var GameMode = /** @class */ (function () {
    /**
     * Create a {@link GameMode} from a key. Key is normalized; unknown keys throw.
     * @param key Game mode key or reasonable variant (case/format-insensitive)
     * @throws Error if the key is not recognized
     */
    function GameMode(key) {
        var normalized = GameMode.normalizeKey(key);
        var def = (0, Helpers_1.getFromRegistry)(normalized, MODE_DEFINITIONS);
        if (!def) {
            throw new Error("Unknown game mode: ".concat(key));
        }
        this.Key = normalized;
        this.DisplayName = def.displayName;
        this.ShortName = def.shortName;
    }
    // Utility helpers
    /** Normalize a mode key (case, separators) to its canonical form. */
    GameMode.normalizeKey = function (key) {
        return (0, Helpers_1.normalizeKey)(key);
    };
    /** Check if a key corresponds to a known game mode. */
    GameMode.exists = function (key) {
        return (0, Helpers_1.hasNormalizedKey)(key, MODE_DEFINITIONS, GameMode.normalizeKey);
    };
    /** Try to get a {@link GameMode} from a key; returns null if unknown. */
    GameMode.tryGet = function (key) {
        return GameMode.exists(key) ? new GameMode(key) : null;
    };
    /**
     * Get all known game modes as {@link GameMode} objects.
     * @returns Readonly array of all modes in the registry
     */
    GameMode.all = function () {
        return (0, Helpers_1.mapRegistry)(MODE_DEFINITIONS, function (k) { return new GameMode(k); });
    };
    Object.defineProperty(GameMode, "CC150", {
        // Static members for each GameModeKey (PascalCase)
        /** 150cc mode. */
        get: function () {
            return new GameMode('150cc');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "CC200", {
        /** 200cc mode. */
        get: function () {
            return new GameMode('200cc');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "RT", {
        /** Regular Tracks. */
        get: function () {
            return new GameMode('rt');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "CT", {
        /** Custom Tracks. */
        get: function () {
            return new GameMode('ct');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "VSRace", {
        /** VS Race. */
        get: function () {
            return new GameMode('vsrace');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "MatchRace", {
        /** Match Race. */
        get: function () {
            return new GameMode('match_race');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "MixedBattle", {
        /** Mixed Battle. */
        get: function () {
            return new GameMode('mixed_battle');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "BalloonBattle", {
        /** Balloon Battle. */
        get: function () {
            return new GameMode('balloon_battle');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "ShineThief", {
        /** Shine Thief. */
        get: function () {
            return new GameMode('shine_thief');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "BobOmbBlast", {
        /** Bob-omb Blast. */
        get: function () {
            return new GameMode('bobomb_blast');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "CoinRunners", {
        /** Coin Runners. */
        get: function () {
            return new GameMode('coin_runners');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "RenegadeRoundup", {
        /** Renegade Roundup. */
        get: function () {
            return new GameMode('renegade_roundup');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GameMode, "Mixed", {
        /** Mixed. */
        get: function () {
            return new GameMode('mixed');
        },
        enumerable: false,
        configurable: true
    });
    return GameMode;
}());
exports.GameMode = GameMode;
