// mkcentral-api/dist/lib/common/Game.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Game = void 0;
var FriendCode_1 = require("../player/FriendCode");
var Platform_1 = require("./Platform");
var Helpers_1 = require("./Helpers");
// Central game registry (single source of truth)
/**
 * Central registry of all known games keyed by {@link GameKey}.
 * This is the single source of truth used to construct {@link Game} instances.
 */
var GAME_DEFINITIONS = Object.freeze({
    mkworld: {
        displayName: 'Mario Kart World',
        shortName: 'MKWORLD',
        platform: 'switch',
        friendCodeType: FriendCode_1.FriendCodeType.Switch,
    },
    mk8dx: {
        displayName: 'Mario Kart 8 Deluxe',
        shortName: 'MK8DX',
        platform: 'switch',
        friendCodeType: FriendCode_1.FriendCodeType.Switch,
    },
    mk7: {
        displayName: 'Mario Kart 7',
        shortName: 'MK7',
        platform: '3ds',
        friendCodeType: FriendCode_1.FriendCodeType.ThreeDS,
    },
    mk8: {
        displayName: 'Mario Kart 8',
        shortName: 'MK8',
        platform: 'wiiu',
        friendCodeType: FriendCode_1.FriendCodeType.NNID,
    },
    mkt: {
        displayName: 'Mario Kart Tour',
        shortName: 'MKT',
        platform: 'mobile',
        friendCodeType: FriendCode_1.FriendCodeType.MKT,
    },
    mkw: {
        displayName: 'Mario Kart Wii',
        shortName: 'MKW',
        platform: 'wii',
        friendCodeType: FriendCode_1.FriendCodeType.MKW,
    },
    smk: {
        displayName: 'Super Mario Kart',
        shortName: 'SMK',
        platform: 'switch',
        friendCodeType: FriendCode_1.FriendCodeType.Switch,
    },
});
/**
 * Immutable game metadata (display/short names, platform, friend-code type).
 * Construct via key or use the provided static getters (e.g., {@link Game.MK8dx}).
 */
var Game = /** @class */ (function () {
    /**
     * Create a {@link Game} from a key. Key is normalized; unknown keys throw.
     * @param key Game key or reasonable variant (case/format-insensitive)
     * @throws Error if the key is not recognized
     */
    function Game(key) {
        var normalized = Game.normalizeKey(key);
        var def = (0, Helpers_1.getFromRegistry)(normalized, GAME_DEFINITIONS);
        if (!def) {
            throw new Error("Unknown game key: ".concat(key));
        }
        this.Key = normalized;
        this.DisplayName = def.displayName;
        this.ShortName = def.shortName;
        this.Platform = new Platform_1.Platform(def.platform);
        this.FriendCodeType = def.friendCodeType;
    }
    // Utility helpers
    /** Normalize a game key (case, separators) to its canonical form. */
    Game.normalizeKey = function (key) {
        return (0, Helpers_1.normalizeKey)(key);
    };
    /** Check if a key corresponds to a known game. */
    Game.exists = function (key) {
        return (0, Helpers_1.hasNormalizedKey)(key, GAME_DEFINITIONS, Game.normalizeKey);
    };
    /** Try to get a {@link Game} from a key; returns null if unknown. */
    Game.tryGet = function (key) {
        return Game.exists(key) ? new Game(key) : null;
    };
    /** Get the {@link FriendCodeType} associated with a game key, if known. */
    Game.getFriendCodeType = function (key) {
        var def = (0, Helpers_1.getFromRegistry)(Game.normalizeKey(key), GAME_DEFINITIONS);
        return def === null || def === void 0 ? void 0 : def.friendCodeType;
    };
    /**
     * Get all known games as {@link Game} objects.
     * @returns Readonly array of all games in the registry
     */
    Game.all = function () {
        return (0, Helpers_1.mapRegistry)(GAME_DEFINITIONS, function (k) { return new Game(k); });
    };
    Object.defineProperty(Game, "MKWorld", {
        // Static members for each GameKey (PascalCase)
        /** Mario Kart World. */
        get: function () {
            return new Game('mkworld');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Game, "MK8dx", {
        /** Mario Kart 8 Deluxe. */
        get: function () {
            return new Game('mk8dx');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Game, "MK7", {
        /** Mario Kart 7. */
        get: function () {
            return new Game('mk7');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Game, "MK8", {
        /** Mario Kart 8. */
        get: function () {
            return new Game('mk8');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Game, "MKT", {
        /** Mario Kart Tour. */
        get: function () {
            return new Game('mkt');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Game, "MKW", {
        /** Mario Kart Wii. */
        get: function () {
            return new Game('mkw');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Game, "SMK", {
        /** Super Mario Kart. */
        get: function () {
            return new Game('smk');
        },
        enumerable: false,
        configurable: true
    });
    return Game;
}());
exports.Game = Game;
