// mkcentral-api/dist/lib/common/Helpers.js
"use strict";
// Common helper utilities shared across models
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapRegistry = exports.getFromRegistry = exports.hasNormalizedKey = exports.normalizeKey = exports.toNullableBoolean = exports.toBoolean = void 0;
/**
 * Coerce a value into a boolean. Undefined/null/unknown values become false.
 */
function toBoolean(v) {
    if (typeof v === "boolean")
        return v;
    if (typeof v === "number")
        return v !== 0;
    if (typeof v === "string")
        return v === "1" || v.toLowerCase() === "true";
    return false;
}
exports.toBoolean = toBoolean;
/**
 * Coerce a value into a boolean or null. Undefined/null/unknown values become null.
 */
function toNullableBoolean(v) {
    if (v === undefined || v === null)
        return null;
    if (typeof v === "boolean")
        return v;
    if (typeof v === "number")
        return v !== 0;
    if (typeof v === "string")
        return v === "1" || v.toLowerCase() === "true";
    return null;
}
exports.toNullableBoolean = toNullableBoolean;
// Generic registry helpers (for Game, Platform, etc.)
/** Normalize a key: trim and lowercase. */
function normalizeKey(key) {
    return (key || '').trim().toLowerCase();
}
exports.normalizeKey = normalizeKey;
/** Check if a (normalized) key exists in the provided registry. */
function hasNormalizedKey(key, registry, normalizer) {
    if (normalizer === void 0) { normalizer = normalizeKey; }
    var k = normalizer(key);
    return k in registry;
}
exports.hasNormalizedKey = hasNormalizedKey;
/** Get a value from a registry by normalized key. */
function getFromRegistry(key, registry, normalizer) {
    if (normalizer === void 0) { normalizer = normalizeKey; }
    var k = normalizer(key);
    return registry[k];
}
exports.getFromRegistry = getFromRegistry;
/** Map over a registry into an array using key and value. */
function mapRegistry(registry, mapper) {
    return Object.keys(registry).map(function (k) {
        var key = k;
        return mapper(key, registry[key]);
    });
}
exports.mapRegistry = mapRegistry;
