// mkcentral-api/dist/lib/common/Platform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Platform = void 0;
var Helpers_1 = require("./Helpers");
var PLATFORM_DEFINITIONS = Object.freeze({
    switch: { key: 'switch', displayName: 'Nintendo Switch' },
    '3ds': { key: '3ds', displayName: 'Nintendo 3DS' },
    wiiu: { key: 'wiiu', displayName: 'Wii U' },
    wii: { key: 'wii', displayName: 'Wii' },
    mobile: { key: 'mobile', displayName: 'Mobile' },
    snes: { key: 'snes', displayName: 'Super Nintendo' },
    unknown: { key: 'unknown', displayName: 'Unknown' },
});
var Platform = /** @class */ (function () {
    function Platform(key) {
        var _a;
        var normalized = Platform.normalizeKey(key);
        var def = (_a = (0, Helpers_1.getFromRegistry)(normalized, PLATFORM_DEFINITIONS)) !== null && _a !== void 0 ? _a : PLATFORM_DEFINITIONS.unknown;
        this.Key = def.key;
        this.DisplayName = def.displayName;
    }
    Platform.normalizeKey = function (key) {
        return (0, Helpers_1.normalizeKey)(key);
    };
    Platform.exists = function (key) {
        return (0, Helpers_1.hasNormalizedKey)(key, PLATFORM_DEFINITIONS, Platform.normalizeKey);
    };
    Platform.tryGet = function (key) {
        return Platform.exists(key) ? new Platform(key) : null;
    };
    Platform.all = function () {
        return (0, Helpers_1.mapRegistry)(PLATFORM_DEFINITIONS, function (k) { return new Platform(k); });
    };
    return Platform;
}());
exports.Platform = Platform;
