// mkcentral-api/dist/lib/common/Roster.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Roster = void 0;
var Game_1 = require("./Game");
var Helpers_1 = require("./Helpers");
var GameMode_1 = require("./GameMode");
// TODO: Consider collapsing Name <-> RosterName | TeamName and Tag <-> RosterTag | TeamTag once usage is audited.
var Roster = /** @class */ (function () {
    /** Construct from a JSON string or a plain object; normalizes snake_case fields to PascalCase with null defaults. */
    function Roster(data) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
        var obj = typeof data === "string" ? JSON.parse(data) : (data !== null && data !== void 0 ? data : {});
        this.Game = obj.game ? Game_1.Game.tryGet(obj.game) : null;
        this.Mode = obj.mode ? GameMode_1.GameMode.tryGet(obj.mode) : null;
        // Player-context
        this.JoinDate = (_a = obj.join_date) !== null && _a !== void 0 ? _a : null;
        this.IsBaggerClause = (_b = obj.is_bagger_clause) !== null && _b !== void 0 ? _b : null;
        // Normalize ids and aliases across shapes (Id is canonical)
        var id = (_c = obj.id) !== null && _c !== void 0 ? _c : obj.roster_id;
        this.Id = id !== null && id !== void 0 ? id : null;
        this.Name = (_d = obj.name) !== null && _d !== void 0 ? _d : null;
        this.Tag = (_e = obj.tag) !== null && _e !== void 0 ? _e : null;
        this.RosterName = (_f = obj.roster_name) !== null && _f !== void 0 ? _f : ((_g = this.Name) !== null && _g !== void 0 ? _g : null);
        this.RosterTag = (_h = obj.roster_tag) !== null && _h !== void 0 ? _h : ((_j = this.Tag) !== null && _j !== void 0 ? _j : null);
        this.TeamId = (_k = obj.team_id) !== null && _k !== void 0 ? _k : null;
        this.TeamName = (_l = obj.team_name) !== null && _l !== void 0 ? _l : null;
        this.TeamTag = (_m = obj.team_tag) !== null && _m !== void 0 ? _m : null;
        this.ApprovalStatus = (_o = obj.approval_status) !== null && _o !== void 0 ? _o : null;
        this.Color = (_q = ((_p = obj.color) !== null && _p !== void 0 ? _p : obj.team_color)) !== null && _q !== void 0 ? _q : null;
        this.CreationDate = (_r = obj.creation_date) !== null && _r !== void 0 ? _r : null;
        this.IsActive =
            obj.is_active === undefined || obj.is_active === null ? null : (0, Helpers_1.toBoolean)(obj.is_active);
        this.IsRecruiting =
            obj.is_recruiting === undefined || obj.is_recruiting === null
                ? null
                : (0, Helpers_1.toBoolean)(obj.is_recruiting);
        // Map nested arrays with dynamic require to avoid circular deps
        var rawPlayers = Array.isArray(obj.players) ? obj.players : [];
        var rawInvites = Array.isArray(obj.invites) ? obj.invites : [];
        var PlayerCtor = rawPlayers.length ? require("../Player").Player : null;
        var InviteCtor = rawInvites.length ? require("../team/Invite").Invite : null;
        this.Players = PlayerCtor ? rawPlayers.map(function (p) { return new PlayerCtor(p); }) : [];
        this.Invites = InviteCtor ? rawInvites.map(function (i) { return new InviteCtor(i); }) : [];
    }
    Object.defineProperty(Roster.prototype, "IsTeamContext", {
        /** True when the instance has team-roster context (e.g., Id/Players/ApprovalStatus present). */
        get: function () {
            return this.Id != null || this.Players.length > 0 || this.ApprovalStatus != null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Roster.prototype, "IsPlayerContext", {
        /** True when the instance has player-roster context (e.g., JoinDate/TeamName/TeamTag present). */
        get: function () {
            return this.Id != null || this.JoinDate != null || this.TeamName != null || this.TeamTag != null;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Load the full roster (team-context) corresponding to this roster using Team.Get(TeamId),
     * then copy team-side fields onto this instance to preserve identity.
     *
     * Throws when TeamId is null.
     */
    Roster.prototype.LoadFromTeam = function () {
        var _a, _b, _c, _d, _e;
        return __awaiter(this, void 0, void 0, function () {
            var TeamMod, team, targetId, match, tag_1, modeKey_1;
            var _this = this;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        if (this.TeamId == null) {
                            throw new Error("Cannot load roster from team without TeamId");
                        }
                        TeamMod = require("../Team");
                        return [4 /*yield*/, TeamMod.Team.Get(this.TeamId)];
                    case 1:
                        team = _f.sent();
                        targetId = (_a = this.Id) !== null && _a !== void 0 ? _a : this.Id;
                        match = targetId != null ? ((_b = team.Rosters.find(function (r) { return r.Id === targetId; })) !== null && _b !== void 0 ? _b : null) : null;
                        if (!match) {
                            tag_1 = (_c = this.Tag) !== null && _c !== void 0 ? _c : this.RosterTag;
                            modeKey_1 = (_d = this.Mode) === null || _d === void 0 ? void 0 : _d.Key;
                            match =
                                (_e = team.Rosters.find(function (r) {
                                    var _a;
                                    return (tag_1 ? r.Tag === tag_1 : true) &&
                                        (modeKey_1 ? (((_a = r.Mode) === null || _a === void 0 ? void 0 : _a.Key) === modeKey_1) : true) &&
                                        (_this.Game && r.Game ? r.Game === _this.Game : true);
                                })) !== null && _e !== void 0 ? _e : null;
                        }
                        if (match) {
                            this.ApprovalStatus = match.ApprovalStatus;
                            this.Color = match.Color;
                            this.CreationDate = match.CreationDate;
                            this.Game = match.Game;
                            this.Id = match.Id;
                            this.IsActive = match.IsActive;
                            this.IsRecruiting = match.IsRecruiting;
                            this.Mode = match.Mode;
                            this.Name = match.Name;
                            this.Tag = match.Tag;
                            this.TeamId = match.TeamId;
                            this.Players = match.Players;
                            this.Invites = match.Invites;
                            // Backfill player-side aliases if missing
                            if (this.RosterName == null)
                                this.RosterName = this.Name;
                            if (this.RosterTag == null)
                                this.RosterTag = this.Tag;
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Load and return the owning Team object using TeamId.
     *
     * Throws when TeamId is null.
     */
    Roster.prototype.LoadTeam = function () {
        return __awaiter(this, void 0, void 0, function () {
            var TeamMod;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.TeamId == null) {
                            throw new Error("Cannot load team without TeamId");
                        }
                        TeamMod = require("../Team");
                        return [4 /*yield*/, TeamMod.Team.Get(this.TeamId)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return Roster;
}());
exports.Roster = Roster;
