// mkcentral-api/dist/lib/team/Invite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Invite = void 0;
var Discord_1 = require("../player/Discord");
var FriendCode_1 = require("../player/FriendCode");
var Invite = /** @class */ (function () {
    function Invite(data) {
        var _a, _b, _c, _d;
        var obj;
        if (typeof data === 'string') {
            obj = JSON.parse(data);
        }
        else {
            obj = data;
        }
        this.CountryCode = (_a = obj.country_code) !== null && _a !== void 0 ? _a : null;
        this.Discord = obj.discord ? new Discord_1.Discord(obj.discord) : null;
        this.FriendCodes = Array.isArray(obj.friend_codes) ? obj.friend_codes.map(function (fc) { return new FriendCode_1.FriendCode(fc); }) : null;
        this.InviteDate = (_b = obj.invite_date) !== null && _b !== void 0 ? _b : null;
        this.IsBaggerClause = obj.is_bagger_clause === undefined || obj.is_bagger_clause === null ? null : (obj.is_bagger_clause === 1 || obj.is_bagger_clause === true || obj.is_bagger_clause === '1');
        this.IsBanned = obj.is_banned === undefined || obj.is_banned === null ? null : (obj.is_banned === 1 || obj.is_banned === true || obj.is_banned === '1');
        this.Name = (_c = obj.name) !== null && _c !== void 0 ? _c : null;
        this.PlayerId = (_d = obj.player_id) !== null && _d !== void 0 ? _d : null;
    }
    return Invite;
}());
exports.Invite = Invite;
