// mkcentral-api/dist/lib/team/Manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Manager = void 0;
var Discord_1 = require("../player/Discord");
var Manager = /** @class */ (function () {
    function Manager(data) {
        var _a, _b, _c, _d;
        var obj;
        if (typeof data === 'string') {
            obj = JSON.parse(data);
        }
        else {
            obj = data;
        }
        this.CountryCode = (_a = obj.country_code) !== null && _a !== void 0 ? _a : null;
        this.Discord = obj.discord ? new Discord_1.Discord(obj.discord) : null;
        this.Id = (_b = obj.id) !== null && _b !== void 0 ? _b : null;
        // Convert 0/1 (or boolean) to boolean, keep null when absent
        this.IsBanned = obj.is_banned === undefined || obj.is_banned === null ? null : (obj.is_banned === 1 || obj.is_banned === true || obj.is_banned === '1');
        this.IsHidden = obj.is_hidden === undefined || obj.is_hidden === null ? null : (obj.is_hidden === 1 || obj.is_hidden === true || obj.is_hidden === '1');
        this.IsShadow = obj.is_shadow === undefined || obj.is_shadow === null ? null : (obj.is_shadow === 1 || obj.is_shadow === true || obj.is_shadow === '1');
        this.JoinDate = (_c = obj.join_date) !== null && _c !== void 0 ? _c : null;
        this.Name = (_d = obj.name) !== null && _d !== void 0 ? _d : null;
    }
    return Manager;
}());
exports.Manager = Manager;
