// mkcentral-api/dist/lib/team/Roster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Roster = void 0;
var Roster_1 = require("../common/Roster");
Object.defineProperty(exports, "Roster", { enumerable: true, get: function () { return Roster_1.Roster; } });
// import { toBoolean } from "../common/Helpers";
// import { Invite } from "./Invite";
// import { Game } from "../common/Game";
// export class Roster {
// 	/** Approval status */
// 	public ApprovalStatus: string;
// 	/** Team color */
// 	public Color: number;
// 	/** Creation date (unix timestamp) */
// 	public CreationDate: number;
// 	/** Game */
// 	public Game: Game;
// 	/** Roster ID */
// 	public Id: number;
// 	/** Whether roster is active */
// 	public IsActive: boolean;
// 	/** Whether roster is recruiting */
// 	public IsRecruiting: boolean;
// 	/** Game mode */
// 	public Mode: string;
// 	/** Roster name */
// 	public Name: string;
// 	/** Roster tag */
// 	public Tag: string;
// 	/** Parent team ID */
// 	public TeamId: number;
// 	/** Players on the roster (roster-embedded Player variant) */
// 	public Players: Player[];
// 	/** Pending invites for the roster */
// 	public Invites: Invite[];
// 	constructor(data: string | object) {
// 		let obj: any;
// 		if (typeof data === "string") {
// 			obj = JSON.parse(data);
// 		} else {
// 			obj = data ?? {};
// 		}
// 		this.ApprovalStatus = obj.approval_status;
// 		this.Color = obj.color;
// 		this.CreationDate = obj.creation_date;
// 		this.Game = new Game(obj.game);
// 		this.Id = obj.id;
// 		this.IsActive = toBoolean(obj.is_active);
// 		this.IsRecruiting = toBoolean(obj.is_recruiting);
// 		this.Mode = obj.mode;
// 		this.Name = obj.name;
// 		this.Tag = obj.tag;
// 		this.TeamId = obj.team_id;
// 		this.Players = Array.isArray(obj.players) ? obj.players.map((p: any) => new Player(p)) : [];
// 		this.Invites = Array.isArray(obj.invites) ? obj.invites.map((i: any) => new Invite(i)) : [];
// 	}
// }
