// mkcentral-api/dist/lib/team/TeamFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamFilter = void 0;
var GameMode_1 = require("../common/GameMode");
var Game_1 = require("../common/Game");
var Language_1 = require("../common/Language");
var TeamFilter = /** @class */ (function () {
    function TeamFilter(init) {
        /**
         * Filter by team name only
         */
        this.Name = null;
        /**
         * Filter by team tag only
         */
        this.Tag = null;
        /**
         * Name or tag search
         */
        this.NameOrTag = null;
        /**
         * Game to filter (e.g., Game.tryGet('mk8dx'))
         */
        this.Game = null;
        /**
         * Game mode to filter (e.g., GameMode.tryGet('150cc'))
         */
        this.Mode = null;
        /**
         * Preferred language to filter (e.g., Language.EnglishTraditional)
         */
        this.Language = null;
        /**
         * Filter recruiting teams
         */
        this.IsRecruiting = null;
        /**
         * Filter historical teams
         */
        this.IsHistorical = null;
        /**
         * Filter active teams
         */
        this.IsActive = null;
        /**
         * Minimum roster size to filter
         */
        this.MinPlayerCount = null;
        /**
         * Sort results by newest registration
         */
        this.SortByNewest = false;
        /**
         * Page number for results
         */
        this.Page = null;
        if (!init)
            return;
        Object.assign(this, init);
        // Normalize Mode when provided as a string
        if (typeof init.Mode === 'string') {
            this.Mode = GameMode_1.GameMode.tryGet(init.Mode);
        }
        // Normalize Game when provided as a string
        if (typeof init.Game === 'string') {
            this.Game = Game_1.Game.tryGet(init.Game);
        }
        // Normalize Language when provided as a string
        if (typeof init.Language === 'string') {
            this.Language = Object.values(Language_1.Language).includes(init.Language)
                ? init.Language
                : null;
        }
    }
    /**
     * Build a query string of URL parameters from the set fields.
     * Undefined/null/false values are omitted. Booleans are serialized as 'true'.
     */
    TeamFilter.prototype.ToQueryString = function () {
        var params = {};
        var add = function (k, v) {
            if (v === null || v === undefined)
                return;
            if (v instanceof GameMode_1.GameMode) {
                params[k] = v.Key;
                return;
            }
            if (v instanceof Game_1.Game) {
                params[k] = v.Key;
                return;
            }
            if (typeof v === 'string') {
                if (v.length === 0)
                    return;
                params[k] = v;
            }
            else if (typeof v === 'number') {
                if (!Number.isFinite(v))
                    return;
                params[k] = String(v);
            }
            else if (typeof v === 'boolean') {
                if (v === false)
                    return;
                params[k] = String(v);
            }
            else {
                params[k] = String(v);
            }
        };
        add('name', this.Name);
        add('tag', this.Tag);
        add('name_or_tag', this.NameOrTag);
        add('game', this.Game);
        add('mode', this.Mode);
        add('language', this.Language);
        add('is_recruiting', this.IsRecruiting);
        add('is_historical', this.IsHistorical);
        add('is_active', this.IsActive);
        add('min_player_count', this.MinPlayerCount);
        add('sort_by_newest', this.SortByNewest);
        add('page', this.Page);
        return Object.keys(params)
            .map(function (k) { return "".concat(encodeURIComponent(k), "=").concat(encodeURIComponent(params[k])); })
            .join('&');
    };
    /**
     * Create a filter with repository defaults
     */
    TeamFilter.Default = function () {
        return new TeamFilter({
            Name: null,
            Tag: null,
            NameOrTag: null,
            Game: null,
            Mode: null,
            Language: null,
            IsRecruiting: null,
            IsHistorical: false,
            IsActive: true,
            MinPlayerCount: 6,
            SortByNewest: false,
            Page: 1,
        });
    };
    return TeamFilter;
}());
exports.TeamFilter = TeamFilter;
