// mkcentral-api/dist/lib/player/FriendCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FriendCodeType = exports.FriendCode = void 0;
var FriendCode = /** @class */ (function () {
    function FriendCode(data) {
        var _a, _b;
        var obj;
        if (typeof data === 'string') {
            obj = JSON.parse(data);
        }
        else {
            obj = data;
        }
        this.Id = obj.id;
        this.Fc = obj.fc;
        this.Type = obj.type;
        this.PlayerId = obj.player_id;
        this.IsVerified = obj.is_verified;
        this.IsPrimary = obj.is_primary;
        this.CreationDate = obj.creation_date;
        this.Description = (_a = obj.description) !== null && _a !== void 0 ? _a : null;
        this.IsActive = (_b = obj.is_active) !== null && _b !== void 0 ? _b : true;
    }
    return FriendCode;
}());
exports.FriendCode = FriendCode;
var FriendCodeType;
(function (FriendCodeType) {
    FriendCodeType["Switch"] = "switch";
    FriendCodeType["NNID"] = "nnid";
    FriendCodeType["ThreeDS"] = "3ds";
    FriendCodeType["MKW"] = "mkw";
    FriendCodeType["MKT"] = "mkt";
})(FriendCodeType = exports.FriendCodeType || (exports.FriendCodeType = {}));
// Map moved to Game.ts (use Game.getFriendCodeType or new Game(key).FriendCodeType)
// export const GameFCMap ... removed.
//     'mkworld': FriendCodeType.Switch,
//     'mk8dx': FriendCodeType.Switch,
//     'mk7': FriendCodeType.ThreeDS,
//     'mk8': FriendCodeType.NNID,
//     'mkt': FriendCodeType.MKT,
//     'mkw': FriendCodeType.MKW,
//     'smk': FriendCodeType.Switch,
// }
