// mkcentral-api/dist/lib/player/PlayerFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerFilter = void 0;
var PlayerFilter = /** @class */ (function () {
    function PlayerFilter(init) {
        /** Filter by player name only */
        this.Name = null;
        /** Filter by player Friend Code only */
        this.FriendCode = null;
        /** Name or friend code search */
        this.NameOrFC = null;
        /** Type of friend code to match */
        this.FCType = null;
        /** Country code to filter */
        this.Country = null;
        /** Filter hidden players */
        this.IsHidden = null;
        /** Filter shadow players */
        this.IsShadow = null;
        /** Filter banned players */
        this.IsBanned = null;
        /** Discord user ID to filter */
        this.DiscordId = null;
        /** Request detailed player data */
        this.Detailed = null;
        /** Page number for results */
        this.Page = null;
        /** Registration ID to filter */
        this.RegistrationId = null;
        /** Only include players with matching friend codes */
        this.MatchingFCsOnly = false;
        /** Include banned, hidden, and shadow players in results */
        this.IncludeShadowPlayers = false;
        /** Sort results by newest registration */
        this.SortByNewest = false;
        /** Filter by connected user presence */
        this.HasConnectedUser = null;
        if (!init)
            return;
        Object.assign(this, init);
    }
    /**
     * Build a query string of URL parameters from the set fields.
     *
     * Undefined/null/false values are omitted. Booleans are serialized as 'true'.
     */
    PlayerFilter.prototype.ToQueryString = function () {
        var params = {};
        var add = function (k, v) {
            if (v === null || v === undefined)
                return;
            if (typeof v === 'string') {
                if (v.length === 0)
                    return;
                params[k] = v;
            }
            else if (typeof v === 'number') {
                if (!Number.isFinite(v))
                    return;
                params[k] = String(v);
            }
            else if (typeof v === 'boolean') {
                if (v === false)
                    return;
                params[k] = String(v);
            }
            else {
                params[k] = String(v);
            }
        };
        add('name', this.Name);
        add('friend_code', this.FriendCode);
        add('name_or_fc', this.NameOrFC);
        add('fc_type', this.FCType);
        add('country', this.Country);
        add('is_hidden', this.IsHidden);
        add('is_shadow', this.IsShadow);
        add('is_banned', this.IsBanned);
        add('discord_id', this.DiscordId);
        add('detailed', this.Detailed);
        add('page', this.Page);
        add('registration_id', this.RegistrationId);
        add('matching_fcs_only', this.MatchingFCsOnly);
        add('include_shadow_players', this.IncludeShadowPlayers);
        add('sort_by_newest', this.SortByNewest);
        add('has_connected_user', this.HasConnectedUser);
        return Object.keys(params)
            .map(function (k) { return "".concat(encodeURIComponent(k), "=").concat(encodeURIComponent(params[k])); })
            .join('&');
    };
    /**
     * Create a filter with repository defaults
     */
    PlayerFilter.Default = function () {
        return new PlayerFilter({
            Name: null,
            FriendCode: null,
            NameOrFC: null,
            FCType: null,
            Country: null,
            IsHidden: false,
            IsShadow: false,
            IsBanned: false,
            DiscordId: null,
            Detailed: true,
            Page: 1,
            RegistrationId: null,
            MatchingFCsOnly: true,
            IncludeShadowPlayers: false,
            SortByNewest: false,
            HasConnectedUser: null,
        });
    };
    return PlayerFilter;
}());
exports.PlayerFilter = PlayerFilter;
