// mkcentral-api/dist/lib/player/UserSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSettings = void 0;
var APIRequest_1 = require("../common/APIRequest");
var UserSettings = /** @class */ (function () {
    function UserSettings(data) {
        var obj;
        if (typeof data === 'string') {
            obj = JSON.parse(data);
        }
        else {
            obj = data;
        }
        this.UserId = obj.user_id;
        this.Avatar = obj.avatar;
        this.AboutMe = obj.about_me;
        this.Language = obj.language;
        this.ColorScheme = obj.color_scheme;
        this.Timezone = obj.timezone;
        this.HideDiscord = obj.hide_discord;
    }
    Object.defineProperty(UserSettings.prototype, "AvatarImage", {
        get: function () {
            var _this = this;
            if (this._avatarImage)
                return Promise.resolve(this._avatarImage);
            if (this._avatarImagePromise)
                return this._avatarImagePromise;
            if (!this.Avatar) {
                return Promise.reject(new Error("No avatar URL available."));
            }
            var url = /^https?:\/\//i.test(this.Avatar) ? this.Avatar : "https://mkcentral.com".concat(this.Avatar);
            this._avatarImagePromise = (0, APIRequest_1.fetchHttpsBuffer)(url)
                .then(function (buf) {
                _this._avatarImage = buf;
                return buf;
            })
                .finally(function () {
                _this._avatarImagePromise = undefined;
            });
            return this._avatarImagePromise;
        },
        enumerable: false,
        configurable: true
    });
    return UserSettings;
}());
exports.UserSettings = UserSettings;
