// mkcentral-api/dist/lib/Players.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Players = void 0;
var node_url_1 = require("node:url");
var APIRequest_1 = require("./common/APIRequest");
var Player_1 = require("./Player");
var PlayerFilter_1 = require("./player/PlayerFilter");
var FriendCode_1 = require("./player/FriendCode");
var CountryCode_1 = require("./common/CountryCode");
/**
 * Players is a typed array of Player items with extra metadata from the API.
 * Uses PlayerFilter for querying.
 */
var Players = /** @class */ (function (_super) {
    __extends(Players, _super);
    function Players(rawData) {
        var _this = _super.call(this) || this;
        _this.Filter = PlayerFilter_1.PlayerFilter.Default();
        if (!rawData) {
            _this.Count = 0;
            return _this;
        }
        var parsed = JSON.parse(rawData);
        var list = Array.isArray(parsed.player_list) ? parsed.player_list : [];
        for (var _i = 0, list_1 = list; _i < list_1.length; _i++) {
            var item = list_1[_i];
            _this.push(new Player_1.Player(item));
        }
        _this.Count = _this.length;
        _this.PlayerCount = typeof parsed.player_count === 'number' ? parsed.player_count : _this.length;
        _this.PageCount = typeof parsed.page_count === 'number' ? parsed.page_count : undefined;
        return _this;
    }
    /** Refresh this instance using its Filter to fetch fresh data. */
    Players.prototype.Load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var uri, result, _a, _i, result_1, p;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        uri = Players.buildURL(this.Filter);
                        _a = Players.bind;
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(uri)];
                    case 1:
                        result = new (_a.apply(Players, [void 0, _b.sent()]))();
                        // Replace contents while preserving this instance
                        this.length = 0;
                        for (_i = 0, result_1 = result; _i < result_1.length; _i++) {
                            p = result_1[_i];
                            this.push(p);
                        }
                        this.Count = result.Count;
                        this.PlayerCount = result.PlayerCount;
                        this.PageCount = result.PageCount;
                        // Ensure pagination state has a defined page
                        if (typeof this.Filter.Page !== 'number' || this.Filter.Page < 1)
                            this.Filter.Page = 1;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Reload the list with the next page, if available.
     * Returns true if the page advanced and data was loaded; false if already at last page.
     */
    Players.prototype.LoadNextPage = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var current;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        current = typeof ((_a = this.Filter) === null || _a === void 0 ? void 0 : _a.Page) === 'number' && this.Filter.Page > 0 ? this.Filter.Page : 1;
                        if (typeof this.PageCount === 'number' && current >= this.PageCount)
                            return [2 /*return*/, false];
                        this.Filter.Page = current + 1;
                        return [4 /*yield*/, this.Load()];
                    case 1:
                        _b.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    /**
     * Reload the list with the previous page, if available.
     * Returns true if the page moved back and data was loaded; false if already at page 1.
     */
    Players.prototype.LoadPreviousPage = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var current;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        current = typeof ((_a = this.Filter) === null || _a === void 0 ? void 0 : _a.Page) === 'number' && this.Filter.Page > 0 ? this.Filter.Page : 1;
                        if (current <= 1)
                            return [2 /*return*/, false];
                        this.Filter.Page = current - 1;
                        return [4 /*yield*/, this.Load()];
                    case 1:
                        _b.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    /**
     * Load all entries from every available page into this instance.
     * Aggregates results across pages for the current filter and replaces the list contents.
     */
    Players.prototype.LoadAllPages = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var base, tempFilter, all, page, pageCount, playerCount, uri, chunk, _b, _i, chunk_1, p, _c, all_1, p;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        base = (_a = this.Filter) !== null && _a !== void 0 ? _a : PlayerFilter_1.PlayerFilter.Default();
                        tempFilter = new PlayerFilter_1.PlayerFilter(base);
                        all = [];
                        page = 1;
                        _d.label = 1;
                    case 1:
                        if (!true) return [3 /*break*/, 3];
                        tempFilter.Page = page;
                        uri = Players.buildURL(tempFilter);
                        _b = Players.bind;
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(uri)];
                    case 2:
                        chunk = new (_b.apply(Players, [void 0, _d.sent()]))();
                        if (page === 1) {
                            pageCount = chunk.PageCount;
                            playerCount = chunk.PlayerCount;
                        }
                        if (chunk.length === 0)
                            return [3 /*break*/, 3];
                        for (_i = 0, chunk_1 = chunk; _i < chunk_1.length; _i++) {
                            p = chunk_1[_i];
                            all.push(p);
                        }
                        if (typeof pageCount === 'number' && page >= pageCount)
                            return [3 /*break*/, 3];
                        page++;
                        return [3 /*break*/, 1];
                    case 3:
                        // Replace contents while preserving identity
                        this.length = 0;
                        for (_c = 0, all_1 = all; _c < all_1.length; _c++) {
                            p = all_1[_c];
                            this.push(p);
                        }
                        this.Count = this.length;
                        this.PlayerCount = typeof playerCount === 'number' ? playerCount : this.length;
                        this.PageCount = pageCount;
                        // After aggregating, normalize page to 1 for the filter
                        this.Filter.Page = 1;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Perform a players query and return a new Players instance
     */
    Players.Get = function (arg) {
        return __awaiter(this, void 0, void 0, function () {
            var filter, s, uri, res, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (typeof arg === 'string') {
                            filter = PlayerFilter_1.PlayerFilter.Default();
                            s = arg.trim();
                            if (s.length > 0)
                                filter.NameOrFC = s;
                        }
                        else {
                            filter = arg instanceof PlayerFilter_1.PlayerFilter ? arg : new PlayerFilter_1.PlayerFilter(arg);
                        }
                        uri = Players.buildURL(filter);
                        _a = Players.bind;
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(uri)];
                    case 1:
                        res = new (_a.apply(Players, [void 0, _b.sent()]))();
                        // Ensure prototype is set correctly (fixes Array subclass issues)
                        Object.setPrototypeOf(res, Players.prototype);
                        if (typeof filter.Page !== 'number' || filter.Page < 1)
                            filter.Page = 1;
                        res.Filter = filter;
                        return [2 /*return*/, res];
                }
            });
        });
    };
    Players.buildURL = function (filter) {
        var uri = new node_url_1.URL('https://mkcentral.com/api/registry/players');
        var qs = filter.ToQueryString();
        if (qs) {
            // The URL constructor doesn't have a nice bulk setter for params; append directly
            uri.search = qs;
        }
        if (!uri.searchParams.has('detailed'))
            uri.searchParams.set('detailed', 'true');
        if (!uri.searchParams.has('matching_fcs_only'))
            uri.searchParams.set('matching_fcs_only', 'true');
        if (!uri.searchParams.has('is_banned'))
            uri.searchParams.set('is_banned', 'false');
        if (!uri.searchParams.has('is_hidden'))
            uri.searchParams.set('is_hidden', 'false');
        if (!uri.searchParams.has('is_shadow'))
            uri.searchParams.set('is_shadow', 'false');
        if (!uri.searchParams.has('page'))
            uri.searchParams.set('page', '1');
        return uri;
    };
    Players.Filter = PlayerFilter_1.PlayerFilter;
    Players.Item = Player_1.Player;
    Players.FriendCode = FriendCode_1.FriendCode;
    Players.CountryCode = CountryCode_1.CountryCode;
    return Players;
}(Array));
exports.Players = Players;
