// mkcentral-api/dist/lib/Player.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Player = exports.PlayerDataKind = void 0;
var CountryCode_1 = require("./common/CountryCode");
var Helpers_1 = require("./common/Helpers");
var APIRequest_1 = require("./common/APIRequest");
var Roster_1 = require("./common/Roster");
var BanInfo_1 = require("./player/BanInfo");
var Discord_1 = require("./player/Discord");
var FriendCode_1 = require("./player/FriendCode");
var NameChanges_1 = require("./player/NameChanges");
var Roles_1 = require("./player/Roles");
var UserSettings_1 = require("./player/UserSettings");
var PlayerDataKind;
(function (PlayerDataKind) {
    PlayerDataKind["Full"] = "full";
    PlayerDataKind["List"] = "list";
    PlayerDataKind["Roster"] = "roster";
})(PlayerDataKind = exports.PlayerDataKind || (exports.PlayerDataKind = {}));
var Player = /** @class */ (function () {
    function Player(data) {
        var _a, _b, _c, _d, _e;
        var obj;
        if (typeof data === "string") {
            obj = JSON.parse(data);
        }
        else {
            obj = data !== null && data !== void 0 ? data : {};
        }
        // Determine which variant we're dealing with
        var looksRoster = ("player_id" in obj && !("id" in obj)) ||
            ("is_leader" in obj) ||
            ("is_manager" in obj) ||
            ("is_bagger_clause" in obj);
        var kind;
        if (looksRoster) {
            kind = PlayerDataKind.Roster;
        }
        else {
            // ToDo: Check key availability for full/list objects
            var fullKeys = ["roles", "rosters", "ban_info", "user_settings", "name_changes", "notes"];
            var hasAllFullKeys = fullKeys.every(function (k) { return k in obj; });
            kind = hasAllFullKeys ? PlayerDataKind.Full : PlayerDataKind.List;
        }
        this.DataKind = kind;
        this.IsFull = kind === PlayerDataKind.Full;
        // Common mappings
        this.Id = (_b = (_a = obj.id) !== null && _a !== void 0 ? _a : obj.player_id) !== null && _b !== void 0 ? _b : null;
        this.Name = (_c = obj.name) !== null && _c !== void 0 ? _c : null;
        this.CountryCode = ((_d = obj.country_code) !== null && _d !== void 0 ? _d : null);
        this.IsHidden = (0, Helpers_1.toNullableBoolean)(obj.is_hidden);
        this.IsShadow = (0, Helpers_1.toNullableBoolean)(obj.is_shadow);
        this.IsBanned = (0, Helpers_1.toNullableBoolean)(obj.is_banned);
        this.JoinDate = (_e = obj.join_date) !== null && _e !== void 0 ? _e : null;
        this.Discord = obj.discord ? new Discord_1.Discord(obj.discord) : null;
        this.FriendCodes = Array.isArray(obj.friend_codes)
            ? obj.friend_codes.map(function (fc) { return new FriendCode_1.FriendCode(fc); })
            : [];
        // Roster-embedded specific flags
        this.IsLeader = (0, Helpers_1.toNullableBoolean)(obj.is_leader);
        this.IsManager = (0, Helpers_1.toNullableBoolean)(obj.is_manager);
        this.IsBaggerClause = (0, Helpers_1.toNullableBoolean)(obj.is_bagger_clause);
        // Full/List specific collections and objects
        this.Rosters = Array.isArray(obj.rosters)
            ? obj.rosters.map(function (r) { return new Roster_1.Roster(r); })
            : [];
        this.BanInfo = obj.ban_info ? new BanInfo_1.BanInfo(obj.ban_info) : null;
        this.UserSettings = obj.user_settings ? new UserSettings_1.UserSettings(obj.user_settings) : null;
        this.NameChanges = Array.isArray(obj.name_changes)
            ? obj.name_changes.map(function (nc) { return new NameChanges_1.NameChange(nc); })
            : [];
        // Notes are always null publicly; avoid circular imports at runtime.
        this.Notes = null;
        // Roles can be numeric IDs, strings, or objects; Role handles all
        this.Roles = Array.isArray(obj.roles)
            ? obj.roles.map(function (r) { return new Roles_1.Role(r); })
            : [];
    }
    /**
     * Fetch a player's full profile by id and return a populated Player instance.
     */
    Player.Get = function (playerId) {
        return __awaiter(this, void 0, void 0, function () {
            var url, json, trimmed, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (playerId == null || Number.isNaN(playerId)) {
                            throw new TypeError("playerId must be a valid number");
                        }
                        url = "https://mkcentral.com/api/registry/players/".concat(encodeURIComponent(playerId));
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(url)];
                    case 2:
                        json = _a.sent();
                        trimmed = typeof json === "string" ? json.trim() : "";
                        if (!trimmed || trimmed === "null") {
                            throw new Error("Player with id ".concat(playerId, " was not found or no data was returned from the API"));
                        }
                        return [2 /*return*/, new Player(json)];
                    case 3:
                        err_1 = _a.sent();
                        // Provide a clearer not-found error while preserving typed HTTPError
                        if (err_1 && err_1.name === "HTTPError" && err_1.statusCode === 404) {
                            err_1.message = "Player with id ".concat(playerId, " was not found (404)");
                        }
                        throw err_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Load this player's full profile into the current instance (no object replacement).
     * If the instance is already full, this is a no-op.
     */
    Player.prototype.Load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var full;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.IsFull)
                            return [2 /*return*/];
                        if (this.Id == null) {
                            throw new Error("Cannot load player without an Id");
                        }
                        return [4 /*yield*/, Player.Get(this.Id)];
                    case 1:
                        full = _a.sent();
                        // Copy properties to preserve object identity
                        this.Id = full.Id;
                        this.Name = full.Name;
                        this.CountryCode = full.CountryCode;
                        this.IsHidden = full.IsHidden;
                        this.IsShadow = full.IsShadow;
                        this.IsBanned = full.IsBanned;
                        this.JoinDate = full.JoinDate;
                        this.Discord = full.Discord;
                        this.FriendCodes = full.FriendCodes;
                        this.Rosters = full.Rosters;
                        this.BanInfo = full.BanInfo;
                        this.UserSettings = full.UserSettings;
                        this.NameChanges = full.NameChanges;
                        this.Notes = full.Notes;
                        this.Roles = full.Roles;
                        // this.IsLeader = full.IsLeader;
                        // this.IsManager = full.IsManager;
                        // this.IsBaggerClause = full.IsBaggerClause;
                        this.DataKind = full.DataKind;
                        this.IsFull = full.IsFull;
                        return [2 /*return*/];
                }
            });
        });
    };
    // Namespaced static exports for convenient access in consumers
    // Example: Player.FriendCode, Player.Role, Player.Roster, etc.
    Player.DataKindEnum = PlayerDataKind;
    Player.FriendCode = FriendCode_1.FriendCode;
    Player.Discord = Discord_1.Discord;
    Player.BanInfo = BanInfo_1.BanInfo;
    Player.NameChange = NameChanges_1.NameChange;
    Player.Role = Roles_1.Role;
    Player.UserSettings = UserSettings_1.UserSettings;
    Player.Roster = Roster_1.Roster;
    Player.CountryCode = CountryCode_1.CountryCode;
    return Player;
}());
exports.Player = Player;
