// mkcentral-api/dist/lib/Team.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Team = exports.TeamDataKind = void 0;
var APIRequest_1 = require("./common/APIRequest");
var Helpers_1 = require("./common/Helpers");
var Roster_1 = require("./common/Roster");
var Manager_1 = require("./team/Manager");
var Approval_1 = require("./team/Approval");
var Language_1 = require("./common/Language");
var Game_1 = require("./common/Game");
var GameMode_1 = require("./common/GameMode");
var TeamDataKind;
(function (TeamDataKind) {
    TeamDataKind["Full"] = "full";
    TeamDataKind["List"] = "list";
})(TeamDataKind = exports.TeamDataKind || (exports.TeamDataKind = {}));
var Team = /** @class */ (function () {
    function Team(data) {
        var _a, _b, _c, _d, _e, _f, _g;
        var obj = typeof data === 'string' ? JSON.parse(data) : (data !== null && data !== void 0 ? data : {});
        // Determine list vs full variant by probing roster-embedded player flags
        // If any roster player carries is_leader/is_manager/is_bagger_clause, treat as Full
        var rosterPlayers = Array.isArray(obj.rosters)
            ? obj.rosters.flatMap(function (r) { return (Array.isArray(r === null || r === void 0 ? void 0 : r.players) ? r.players : []); })
            : [];
        var hasRosterEmbeddedFlags = rosterPlayers.some(function (p) {
            return p &&
                (typeof p === 'object') &&
                ('is_leader' in p || 'is_manager' in p || 'is_bagger_clause' in p);
        });
        this.DataKind = hasRosterEmbeddedFlags ? TeamDataKind.Full : TeamDataKind.List;
        this.IsFull = this.DataKind === TeamDataKind.Full;
        this.Id = (_a = obj.id) !== null && _a !== void 0 ? _a : null;
        this.Name = (_b = obj.name) !== null && _b !== void 0 ? _b : null;
        this.Tag = (_c = obj.tag) !== null && _c !== void 0 ? _c : null;
        var approvalRaw = typeof obj.approval_status === 'string' ? obj.approval_status.toLowerCase() : null;
        this.ApprovalStatus =
            approvalRaw === Approval_1.Approval.Approved ||
                approvalRaw === Approval_1.Approval.Pending ||
                approvalRaw === Approval_1.Approval.Denied
                ? approvalRaw
                : null;
        this.Color = (_d = obj.color) !== null && _d !== void 0 ? _d : null;
        this.CreationDate = (_e = obj.creation_date) !== null && _e !== void 0 ? _e : null;
        this.Description = (_f = obj.description) !== null && _f !== void 0 ? _f : null;
        // Keep null when absent; otherwise coerce 0/1, "0"/"1", true/false
        this.IsHistorical = obj.is_historical === undefined || obj.is_historical === null
            ? null
            : (0, Helpers_1.toBoolean)(obj.is_historical);
        this.Language = (_g = obj.language) !== null && _g !== void 0 ? _g : null;
        this.Logo = Team.normalizeLogo(obj.logo);
        this.Managers = Array.isArray(obj.managers) ? obj.managers.map(function (m) { return new Manager_1.Manager(m); }) : [];
        this.Rosters = Array.isArray(obj.rosters) ? obj.rosters.map(function (r) { return new Roster_1.Roster(r); }) : [];
    }
    Team.normalizeLogo = function (v) {
        var s = typeof v === 'string' ? v : null;
        if (!s)
            return null;
        if (s.startsWith('http://'))
            return 'https://' + s.slice('http://'.length);
        if (s.startsWith('https://'))
            return s;
        if (s.startsWith('/'))
            return "https://mkcentral.com".concat(s);
        return s;
    };
    /** Fetch a team by ID and return a populated Team instance. */
    Team.Get = function (teamId) {
        return __awaiter(this, void 0, void 0, function () {
            var url, json, trimmed, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (teamId == null || Number.isNaN(teamId)) {
                            throw new TypeError('teamId must be a valid number');
                        }
                        url = "https://mkcentral.com/api/registry/teams/".concat(encodeURIComponent(teamId));
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(url)];
                    case 2:
                        json = _a.sent();
                        trimmed = typeof json === 'string' ? json.trim() : '';
                        if (!trimmed || trimmed === 'null') {
                            throw new Error("Team with id ".concat(teamId, " was not found or no data was returned from the API"));
                        }
                        return [2 /*return*/, new Team(json)];
                    case 3:
                        err_1 = _a.sent();
                        if (err_1 && err_1.name === 'HTTPError' && err_1.statusCode === 404) {
                            err_1.message = "Team with id ".concat(teamId, " was not found (404)");
                        }
                        throw err_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /** Load this team's full profile into the current instance (no object replacement). */
    Team.prototype.Load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var full;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.IsFull)
                            return [2 /*return*/];
                        if (this.Id == null) {
                            throw new Error('Cannot load team without an Id');
                        }
                        return [4 /*yield*/, Team.Get(this.Id)];
                    case 1:
                        full = _a.sent();
                        // Copy properties to preserve object identity
                        this.Id = full.Id;
                        this.Name = full.Name;
                        this.Tag = full.Tag;
                        this.ApprovalStatus = full.ApprovalStatus;
                        this.Color = full.Color;
                        this.CreationDate = full.CreationDate;
                        this.Description = full.Description;
                        this.IsHistorical = full.IsHistorical;
                        this.Language = full.Language;
                        this.Logo = full.Logo;
                        this.Managers = full.Managers;
                        this.Rosters = full.Rosters;
                        this.DataKind = full.DataKind;
                        this.IsFull = full.IsFull;
                        return [2 /*return*/];
                }
            });
        });
    };
    // Namespaced static exports for convenient access
    Team.Manager = Manager_1.Manager;
    Team.Approval = Approval_1.Approval;
    Team.Roster = Roster_1.Roster;
    Team.Language = Language_1.Language;
    Team.Game = Game_1.Game;
    Team.GameMode = GameMode_1.GameMode;
    return Team;
}());
exports.Team = Team;
