// mkcentral-api/dist/lib/Teams.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Teams = void 0;
var node_url_1 = require("node:url");
var APIRequest_1 = require("./common/APIRequest");
var Team_1 = require("./Team");
var TeamFilter_1 = require("./team/TeamFilter");
var Game_1 = require("./common/Game");
var GameMode_1 = require("./common/GameMode");
var Language_1 = require("./common/Language");
/**
 * Teams is a typed array of Team items with extra metadata from the API.
 * Mirrors the structure and behavior of Players.
 */
var Teams = /** @class */ (function (_super) {
    __extends(Teams, _super);
    function Teams(rawData) {
        var _this = _super.call(this) || this;
        _this.Filter = TeamFilter_1.TeamFilter.Default();
        if (!rawData) {
            _this.Count = 0;
            return _this;
        }
        var parsed = rawData ? JSON.parse(rawData) : null;
        // Support multiple possible shapes: team_list (preferred), data (legacy), or teams
        var list = Array.isArray(parsed === null || parsed === void 0 ? void 0 : parsed.team_list)
            ? parsed.team_list
            : Array.isArray(parsed === null || parsed === void 0 ? void 0 : parsed.data)
                ? parsed.data
                : Array.isArray(parsed === null || parsed === void 0 ? void 0 : parsed.teams)
                    ? parsed.teams
                    : [];
        for (var _i = 0, list_1 = list; _i < list_1.length; _i++) {
            var item = list_1[_i];
            _this.push(new Team_1.Team(item));
        }
        _this.Count = _this.length;
        var countFallback = typeof (parsed === null || parsed === void 0 ? void 0 : parsed.count) === 'number' ? parsed.count : _this.length;
        _this.TeamCount = typeof (parsed === null || parsed === void 0 ? void 0 : parsed.team_count) === 'number' ? parsed.team_count : countFallback;
        _this.PageCount = typeof (parsed === null || parsed === void 0 ? void 0 : parsed.page_count) === 'number' ? parsed.page_count : undefined;
        return _this;
    }
    /** Refresh this instance using its Filter to fetch fresh data. */
    Teams.prototype.Load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var uri, result, _a, _i, result_1, t;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        uri = Teams.buildURL(this.Filter);
                        _a = Teams.bind;
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(uri)];
                    case 1:
                        result = new (_a.apply(Teams, [void 0, _b.sent()]))();
                        // Replace contents while preserving this instance
                        this.length = 0;
                        for (_i = 0, result_1 = result; _i < result_1.length; _i++) {
                            t = result_1[_i];
                            this.push(t);
                        }
                        this.Count = result.Count;
                        this.TeamCount = result.TeamCount;
                        this.PageCount = result.PageCount;
                        // Ensure pagination state has a defined page
                        if (typeof this.Filter.Page !== 'number' || this.Filter.Page < 1)
                            this.Filter.Page = 1;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Reload the list with the next page, if available.
     * Returns true if the page advanced and data was loaded; false if already at last page.
     */
    Teams.prototype.LoadNextPage = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var current;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        current = typeof ((_a = this.Filter) === null || _a === void 0 ? void 0 : _a.Page) === 'number' && this.Filter.Page > 0 ? this.Filter.Page : 1;
                        if (typeof this.PageCount === 'number' && current >= this.PageCount)
                            return [2 /*return*/, false];
                        this.Filter.Page = current + 1;
                        return [4 /*yield*/, this.Load()];
                    case 1:
                        _b.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    /**
     * Reload the list with the previous page, if available.
     * Returns true if the page moved back and data was loaded; false if already at page 1.
     */
    Teams.prototype.LoadPreviousPage = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var current;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        current = typeof ((_a = this.Filter) === null || _a === void 0 ? void 0 : _a.Page) === 'number' && this.Filter.Page > 0 ? this.Filter.Page : 1;
                        if (current <= 1)
                            return [2 /*return*/, false];
                        this.Filter.Page = current - 1;
                        return [4 /*yield*/, this.Load()];
                    case 1:
                        _b.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    /**
     * Load all entries from every available page into this instance.
     * Aggregates results across pages for the current filter and replaces the list contents.
     */
    Teams.prototype.LoadAllPages = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var base, tempFilter, all, page, pageCount, teamCount, uri, chunk, _b, _i, chunk_1, t, _c, all_1, t;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        base = (_a = this.Filter) !== null && _a !== void 0 ? _a : TeamFilter_1.TeamFilter.Default();
                        tempFilter = new TeamFilter_1.TeamFilter(base);
                        all = [];
                        page = 1;
                        _d.label = 1;
                    case 1:
                        if (!true) return [3 /*break*/, 3];
                        tempFilter.Page = page;
                        uri = Teams.buildURL(tempFilter);
                        _b = Teams.bind;
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(uri)];
                    case 2:
                        chunk = new (_b.apply(Teams, [void 0, _d.sent()]))();
                        if (page === 1) {
                            pageCount = chunk.PageCount;
                            teamCount = chunk.TeamCount;
                        }
                        if (chunk.length === 0)
                            return [3 /*break*/, 3];
                        for (_i = 0, chunk_1 = chunk; _i < chunk_1.length; _i++) {
                            t = chunk_1[_i];
                            all.push(t);
                        }
                        if (typeof pageCount === 'number' && page >= pageCount)
                            return [3 /*break*/, 3];
                        page++;
                        return [3 /*break*/, 1];
                    case 3:
                        // Replace contents while preserving identity
                        this.length = 0;
                        for (_c = 0, all_1 = all; _c < all_1.length; _c++) {
                            t = all_1[_c];
                            this.push(t);
                        }
                        this.Count = this.length;
                        this.TeamCount = typeof teamCount === 'number' ? teamCount : this.length;
                        this.PageCount = pageCount;
                        // After aggregating, normalize page to 1 for the filter
                        this.Filter.Page = 1;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Perform a teams query and return a new Teams instance.
     * - If arg is a string, it is used as the 'name_or_tag' query.
     * - If arg is a filter, use its fields.
     */
    Teams.Get = function (arg) {
        return __awaiter(this, void 0, void 0, function () {
            var filter, s, uri, res, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (typeof arg === 'string') {
                            filter = TeamFilter_1.TeamFilter.Default();
                            s = arg.trim();
                            if (s.length > 0)
                                filter.NameOrTag = s;
                        }
                        else {
                            filter = arg instanceof TeamFilter_1.TeamFilter ? arg : new TeamFilter_1.TeamFilter(arg);
                        }
                        uri = Teams.buildURL(filter);
                        _a = Teams.bind;
                        return [4 /*yield*/, (0, APIRequest_1.fetchHttpsContent)(uri)];
                    case 1:
                        res = new (_a.apply(Teams, [void 0, _b.sent()]))();
                        // Ensure prototype is set correctly (fixes Array subclass issues)
                        Object.setPrototypeOf(res, Teams.prototype);
                        if (typeof filter.Page !== 'number' || filter.Page < 1)
                            filter.Page = 1;
                        res.Filter = filter;
                        return [2 /*return*/, res];
                }
            });
        });
    };
    Teams.buildURL = function (filter) {
        var uri = new node_url_1.URL('https://mkcentral.com/api/registry/teams');
        var qs = filter.ToQueryString();
        if (qs)
            uri.search = qs;
        if (!uri.searchParams.has('is_historical'))
            uri.searchParams.set('is_historical', 'false');
        if (!uri.searchParams.has('is_active'))
            uri.searchParams.set('is_active', 'true');
        if (!uri.searchParams.has('sort_by_newest'))
            uri.searchParams.set('sort_by_newest', 'false');
        if (!uri.searchParams.has('min_player_count'))
            uri.searchParams.set('min_player_count', '6');
        if (!uri.searchParams.has('page'))
            uri.searchParams.set('page', '1');
        return uri;
    };
    Teams.Filter = TeamFilter_1.TeamFilter;
    Teams.Item = Team_1.Team;
    Teams.Game = Game_1.Game;
    Teams.GameMode = GameMode_1.GameMode;
    Teams.Language = Language_1.Language;
    return Teams;
}(Array));
exports.Teams = Teams;
