// mkcentral-api/dist/examples.js
"use strict";
/* tslint:disable no-console */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runExamples = exports.errorHandlingExamples = exports.combinedExamples = exports.teamExamples = exports.playerExamples = void 0;
var index_1 = require("./index");
// ============================================
// PLAYER EXAMPLES
// ============================================
function playerExamples() {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var searchResults, player, filter, filteredPlayers, partialPlayers, partial, paginatedFilter, paginatedResults, hasNextPage, allPagesFilter, allResults, statusCheck, detailedPlayer, _i, _b, fc, _c, _d, role, refreshExample, initialCount;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    console.log('=== PLAYER EXAMPLES ===\n');
                    // Example 1: Search for players by name
                    console.log('1. Search players by name:');
                    return [4 /*yield*/, index_1.Players.Get('Rollo')];
                case 1:
                    searchResults = _e.sent();
                    console.log("Found ".concat(searchResults.Count, " players matching \"Rollo\""));
                    if (searchResults.length > 0) {
                        console.log("First result: ".concat(searchResults[0].Name, " (ID: ").concat(searchResults[0].Id, ")"));
                    }
                    // Example 2: Get a specific player by ID
                    console.log('\n2. Get player by ID:');
                    return [4 /*yield*/, index_1.Player.Get(1655)];
                case 2:
                    player = _e.sent();
                    console.log("Player: ".concat(player.Name));
                    console.log("Country: ".concat(player.CountryCode));
                    console.log("Discord: ".concat((_a = player.Discord) === null || _a === void 0 ? void 0 : _a.Username));
                    console.log("Friend Codes: ".concat(player.FriendCodes.length));
                    console.log("Rosters: ".concat(player.Rosters.length));
                    // Example 3: Advanced filtering with PlayerFilter
                    console.log('\n3. Filter players by country and game:');
                    filter = new index_1.Players.Filter({
                        Country: index_1.Players.CountryCode.Germany,
                        NameOrFC: 'Alex',
                        Page: 1
                    });
                    return [4 /*yield*/, index_1.Players.Get(filter)];
                case 3:
                    filteredPlayers = _e.sent();
                    console.log("Found ".concat(filteredPlayers.PlayerCount, " total German players matching \"Alex\""));
                    console.log("Showing page 1 of ".concat(filteredPlayers.PageCount));
                    // Example 4: Load partial player data to full
                    console.log('\n4. Load partial player to full:');
                    console.log(' IsFull has some issues, still...');
                    return [4 /*yield*/, index_1.Players.Get('Alex')];
                case 4:
                    partialPlayers = _e.sent();
                    if (!(partialPlayers.length > 0)) return [3 /*break*/, 6];
                    partial = partialPlayers[0];
                    console.log("Before Load - IsFull: ".concat(partial.IsFull));
                    return [4 /*yield*/, partial.Load()];
                case 5:
                    _e.sent();
                    console.log("After Load - IsFull: ".concat(partial.IsFull));
                    console.log("Loaded ".concat(partial.Rosters.length, " rosters"));
                    _e.label = 6;
                case 6:
                    // Example 5: Pagination through player results
                    console.log('\n5. Pagination example:');
                    paginatedFilter = new index_1.Players.Filter({ NameOrFC: 'Anna', Page: 1 });
                    return [4 /*yield*/, index_1.Players.Get(paginatedFilter)];
                case 7:
                    paginatedResults = _e.sent();
                    console.log("Page 1: ".concat(paginatedResults.length, " players"));
                    if (!(typeof paginatedResults.LoadNextPage === 'function')) return [3 /*break*/, 9];
                    return [4 /*yield*/, paginatedResults.LoadNextPage()];
                case 8:
                    hasNextPage = _e.sent();
                    if (hasNextPage) {
                        console.log("Page 2: ".concat(paginatedResults.length, " players"));
                    }
                    return [3 /*break*/, 10];
                case 9:
                    console.log('LoadNextPage method not found - skipping pagination test');
                    _e.label = 10;
                case 10:
                    // Example 6: Load all pages at once
                    console.log('\n6. Load all pages:');
                    allPagesFilter = new index_1.Players.Filter({ Country: index_1.Players.CountryCode.Chile });
                    return [4 /*yield*/, index_1.Players.Get(allPagesFilter)];
                case 11:
                    allResults = _e.sent();
                    console.log("Initial page has ".concat(allResults.length, " players"));
                    return [4 /*yield*/, allResults.LoadAllPages()];
                case 12:
                    _e.sent();
                    console.log("All pages loaded: ".concat(allResults.length, " total Chilean players"));
                    // Example 7: Check player status flags
                    console.log('\n7. Check player status:');
                    return [4 /*yield*/, index_1.Player.Get(183)];
                case 13:
                    statusCheck = _e.sent();
                    console.log("Player: ".concat(statusCheck.Name));
                    console.log("Is Banned: ".concat(statusCheck.IsBanned));
                    console.log("Is Shadow: ".concat(statusCheck.IsShadow));
                    console.log("Is Hidden: ".concat(statusCheck.IsHidden));
                    // Example 8: Access friend codes and roles
                    console.log('\n8. Friend codes and roles:');
                    return [4 /*yield*/, index_1.Player.Get(1655)];
                case 14:
                    detailedPlayer = _e.sent();
                    for (_i = 0, _b = detailedPlayer.FriendCodes; _i < _b.length; _i++) {
                        fc = _b[_i];
                        console.log("".concat(fc.Type, ": ").concat(fc.Fc, " (Verified: ").concat(fc.IsVerified, ")"));
                    }
                    for (_c = 0, _d = detailedPlayer.Roles; _c < _d.length; _c++) {
                        role = _d[_c];
                        console.log("Role: ".concat(role.Name));
                    }
                    // Example 9: Instance refresh with Load()
                    console.log('\n9. Refresh instance data:');
                    return [4 /*yield*/, index_1.Players.Get('Anna')];
                case 15:
                    refreshExample = _e.sent();
                    initialCount = refreshExample.Count;
                    console.log("Initial count: ".concat(initialCount));
                    // Simulate filter change
                    refreshExample.Filter.Page = 2;
                    return [4 /*yield*/, refreshExample.Load()];
                case 16:
                    _e.sent();
                    console.log("After refresh: ".concat(refreshExample.Count, " players on page ").concat(refreshExample.Filter.Page));
                    return [2 /*return*/];
            }
        });
    });
}
exports.playerExamples = playerExamples;
// ============================================
// TEAM EXAMPLES
// ============================================
function teamExamples() {
    var _a, _b, _c, _d, _e;
    return __awaiter(this, void 0, void 0, function () {
        var teamSearch, firstTeam, team, teamFilter, filteredTeams, partialTeams, partialTeam, _i, _f, roster, teamPages, hasNext, hasPrev, detailedTeam, _g, _h, manager, allTeamFilter, allTeams, rosterTeam, _j, _k, roster, _l, _m, player, flags, flagStr;
        return __generator(this, function (_o) {
            switch (_o.label) {
                case 0:
                    console.log('\n=== TEAM EXAMPLES ===\n');
                    // Example 1: Search teams by name or tag
                    console.log('1. Search teams by name/tag:');
                    return [4 /*yield*/, index_1.Teams.Get('HD')];
                case 1:
                    teamSearch = _o.sent();
                    console.log("Found ".concat(teamSearch.Count, " teams matching \"HD\""));
                    if (teamSearch.length > 0) {
                        firstTeam = teamSearch[0];
                        console.log("First result: ".concat(firstTeam.Name, " [").concat(firstTeam.Tag, "]"));
                    }
                    // Example 2: Get specific team by ID
                    console.log('\n2. Get team by ID:');
                    return [4 /*yield*/, index_1.Team.Get(35)];
                case 2:
                    team = _o.sent();
                    console.log("Team: ".concat(team.Name, " [").concat(team.Tag, "]"));
                    console.log("Color: #".concat((_a = team.Color) === null || _a === void 0 ? void 0 : _a.toString(16).padStart(6, '0')));
                    console.log("Language: ".concat(team.Language));
                    console.log("Managers: ".concat(team.Managers.length));
                    console.log("Rosters: ".concat(team.Rosters.length));
                    // Example 3: Filter teams with TeamFilter
                    console.log('\n3. Filter teams:');
                    teamFilter = new index_1.Teams.Filter({
                        NameOrTag: 'Racing',
                        Game: index_1.Teams.Game.MK8dx,
                        Mode: index_1.Teams.GameMode.CC150,
                        Language: index_1.Teams.Language.EnglishSimplified,
                        Page: 1
                    });
                    return [4 /*yield*/, index_1.Teams.Get(teamFilter)];
                case 3:
                    filteredTeams = _o.sent();
                    console.log("Found ".concat(filteredTeams.TeamCount, " total teams matching criteria"));
                    console.log("Page 1 shows ".concat(filteredTeams.Count, " teams"));
                    // Example 4: Load partial team to full
                    console.log('\n4. Load partial team to full:');
                    return [4 /*yield*/, index_1.Teams.Get('Test')];
                case 4:
                    partialTeams = _o.sent();
                    if (!(partialTeams.length > 0)) return [3 /*break*/, 6];
                    partialTeam = partialTeams[0];
                    console.log("Before Load - IsFull: ".concat(partialTeam.IsFull));
                    return [4 /*yield*/, partialTeam.Load()];
                case 5:
                    _o.sent();
                    console.log("After Load - IsFull: ".concat(partialTeam.IsFull));
                    // Access roster players
                    for (_i = 0, _f = partialTeam.Rosters; _i < _f.length; _i++) {
                        roster = _f[_i];
                        console.log("Roster: ".concat((_b = roster.Game) === null || _b === void 0 ? void 0 : _b.DisplayName, " ").concat(roster.Mode));
                        console.log("Players: ".concat((_d = (_c = roster.Players) === null || _c === void 0 ? void 0 : _c.length) !== null && _d !== void 0 ? _d : 0));
                    }
                    _o.label = 6;
                case 6:
                    // Example 5: Team pagination
                    console.log('\n5. Team pagination:');
                    return [4 /*yield*/, index_1.Teams.Get(new index_1.Teams.Filter({ Page: 1 }))];
                case 7:
                    teamPages = _o.sent();
                    console.log("Total teams: ".concat(teamPages.TeamCount));
                    console.log("Total pages: ".concat(teamPages.PageCount));
                    console.log("Current page has ".concat(teamPages.Count, " teams"));
                    if (!(typeof teamPages.LoadNextPage === 'function')) return [3 /*break*/, 10];
                    return [4 /*yield*/, teamPages.LoadNextPage()];
                case 8:
                    hasNext = _o.sent();
                    if (hasNext) {
                        console.log("Page 2 has ".concat(teamPages.Count, " teams"));
                    }
                    return [4 /*yield*/, teamPages.LoadPreviousPage()];
                case 9:
                    hasPrev = _o.sent();
                    if (hasPrev) {
                        console.log("Back to page 1: ".concat(teamPages.Count, " teams"));
                    }
                    return [3 /*break*/, 11];
                case 10:
                    console.log('Team pagination methods not found - skipping pagination test');
                    _o.label = 11;
                case 11:
                    // Example 6: Access team details
                    console.log('\n6. Team details:');
                    return [4 /*yield*/, index_1.Team.Get(35)];
                case 12:
                    detailedTeam = _o.sent();
                    console.log("Team: ".concat(detailedTeam.Name));
                    console.log("Approval Status: ".concat(detailedTeam.ApprovalStatus));
                    console.log("Is Historical: ".concat(detailedTeam.IsHistorical));
                    console.log("Created: ".concat(new Date(detailedTeam.CreationDate * 1000).toLocaleDateString()));
                    // List managers
                    for (_g = 0, _h = detailedTeam.Managers; _g < _h.length; _g++) {
                        manager = _h[_g];
                        console.log("Manager: ".concat(manager.Name, " (ID: ").concat(manager.Id, ")"));
                    }
                    // Example 7: Load all team pages
                    console.log('\n7. Load all team pages:');
                    allTeamFilter = new index_1.Teams.Filter({
                        Game: index_1.Teams.Game.MK8dx,
                        Mode: index_1.Teams.GameMode.CC200
                    });
                    return [4 /*yield*/, index_1.Teams.Get(allTeamFilter)];
                case 13:
                    allTeams = _o.sent();
                    console.log("Initial load: ".concat(allTeams.Count, " teams"));
                    return [4 /*yield*/, allTeams.LoadAllPages()];
                case 14:
                    _o.sent();
                    console.log("All pages loaded: ".concat(allTeams.Count, " total 200cc teams"));
                    // Example 8: Access roster information
                    console.log('\n8. Roster information:');
                    return [4 /*yield*/, index_1.Team.Get(35)];
                case 15:
                    rosterTeam = _o.sent();
                    for (_j = 0, _k = rosterTeam.Rosters; _j < _k.length; _j++) {
                        roster = _k[_j];
                        console.log("\nRoster: ".concat((_e = roster.Game) === null || _e === void 0 ? void 0 : _e.DisplayName, " ").concat(roster.Mode));
                        console.log("Is Recruiting: ".concat(roster.IsRecruiting));
                        if (roster.Players && roster.Players.length > 0) {
                            console.log("Players (".concat(roster.Players.length, "):"));
                            for (_l = 0, _m = roster.Players.slice(0, 3); _l < _m.length; _l++) { // Show first 3
                                player = _m[_l];
                                flags = [];
                                if (player.IsLeader)
                                    flags.push('Leader');
                                if (player.IsManager)
                                    flags.push('Manager');
                                if (player.IsBaggerClause)
                                    flags.push('Bagger Clause');
                                flagStr = flags.length > 0 ? " [".concat(flags.join(', '), "]") : '';
                                console.log("  - ".concat(player.Name).concat(flagStr));
                            }
                        }
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.teamExamples = teamExamples;
// ============================================
// COMBINED EXAMPLES
// ============================================
function combinedExamples() {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function () {
        var player, _i, _c, roster, team, _d, _e, roster, _f, _g, member, roleFlags, prefix, countries, _h, countries_1, country, filter, results;
        return __generator(this, function (_j) {
            switch (_j.label) {
                case 0:
                    console.log('\n=== COMBINED EXAMPLES ===\n');
                    // Example 1: Find player's teams
                    console.log('1. Find all teams for a player:');
                    return [4 /*yield*/, index_1.Player.Get(183)];
                case 1:
                    player = _j.sent();
                    console.log("Player: ".concat(player.Name));
                    console.log("Teams:");
                    for (_i = 0, _c = player.Rosters; _i < _c.length; _i++) {
                        roster = _c[_i];
                        console.log("  - ".concat(roster.TeamName, " [").concat(roster.TeamTag, "] - ").concat((_a = roster.Game) === null || _a === void 0 ? void 0 : _a.DisplayName, " ").concat(roster.Mode));
                    }
                    // Example 2: Find team members
                    console.log('\n2. Find all members of a team:');
                    return [4 /*yield*/, index_1.Team.Get(35)];
                case 2:
                    team = _j.sent();
                    console.log("Team: ".concat(team.Name));
                    for (_d = 0, _e = team.Rosters; _d < _e.length; _d++) {
                        roster = _e[_d];
                        console.log("\n".concat((_b = roster.Game) === null || _b === void 0 ? void 0 : _b.DisplayName, " ").concat(roster.Mode, " Roster:"));
                        if (roster.Players) {
                            for (_f = 0, _g = roster.Players; _f < _g.length; _f++) {
                                member = _g[_f];
                                roleFlags = [];
                                if (member.IsLeader)
                                    roleFlags.push('★ ');
                                if (member.IsManager)
                                    roleFlags.push('♔ ');
                                prefix = roleFlags.length > 0 ? roleFlags.join('') + ' ' : '';
                                console.log("  ".concat(prefix).concat(member.Name));
                            }
                        }
                    }
                    // Example 3: Country statistics
                    console.log('\n3. Players by country:');
                    countries = [index_1.Players.CountryCode.UnitedStates, index_1.Players.CountryCode.Germany, index_1.Players.CountryCode.Japan];
                    _h = 0, countries_1 = countries;
                    _j.label = 3;
                case 3:
                    if (!(_h < countries_1.length)) return [3 /*break*/, 6];
                    country = countries_1[_h];
                    filter = new index_1.Players.Filter({ Country: country, Page: 1 });
                    return [4 /*yield*/, index_1.Players.Get(filter)];
                case 4:
                    results = _j.sent();
                    console.log("".concat(country, ": ").concat(results.PlayerCount, " players"));
                    _j.label = 5;
                case 5:
                    _h++;
                    return [3 /*break*/, 3];
                case 6: return [2 /*return*/];
            }
        });
    });
}
exports.combinedExamples = combinedExamples;
// ============================================
// ERROR HANDLING EXAMPLES
// ============================================
function errorHandlingExamples() {
    return __awaiter(this, void 0, void 0, function () {
        var error_1, error_2, players, error_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    console.log('\n=== ERROR HANDLING EXAMPLES ===\n');
                    // Example 1: Handle invalid player ID
                    console.log('1. Invalid player ID:');
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, index_1.Player.Get(99999999)];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 4];
                case 3:
                    error_1 = _a.sent();
                    console.log("Error: ".concat(error_1.message));
                    return [3 /*break*/, 4];
                case 4:
                    // Example 2: Handle invalid team ID
                    console.log('\n2. Invalid team ID:');
                    _a.label = 5;
                case 5:
                    _a.trys.push([5, 7, , 8]);
                    return [4 /*yield*/, index_1.Team.Get(-1)];
                case 6:
                    _a.sent();
                    return [3 /*break*/, 8];
                case 7:
                    error_2 = _a.sent();
                    console.log("Error: ".concat(error_2.message));
                    return [3 /*break*/, 8];
                case 8:
                    // Example 3: Handle network errors (would need network to be down)
                    console.log('\n3. Network error handling:');
                    _a.label = 9;
                case 9:
                    _a.trys.push([9, 11, , 12]);
                    return [4 /*yield*/, index_1.Players.Get('test')];
                case 10:
                    players = _a.sent();
                    console.log("Success: Found ".concat(players.Count, " players"));
                    return [3 /*break*/, 12];
                case 11:
                    error_3 = _a.sent();
                    if (error_3.name === 'NetworkError') {
                        console.log('Network error occurred');
                    }
                    else if (error_3.name === 'HTTPError') {
                        console.log("HTTP error: ".concat(error_3.message));
                    }
                    else {
                        console.log("Unexpected error: ".concat(error_3.message));
                    }
                    return [3 /*break*/, 12];
                case 12: return [2 /*return*/];
            }
        });
    });
}
exports.errorHandlingExamples = errorHandlingExamples;
// ============================================
// MAIN EXECUTION
// ============================================
function runExamples() {
    return __awaiter(this, void 0, void 0, function () {
        var error_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 5, , 6]);
                    return [4 /*yield*/, playerExamples()];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, teamExamples()];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, combinedExamples()];
                case 3:
                    _a.sent();
                    return [4 /*yield*/, errorHandlingExamples()];
                case 4:
                    _a.sent();
                    return [3 /*break*/, 6];
                case 5:
                    error_4 = _a.sent();
                    console.error('Example execution failed:', error_4);
                    return [3 /*break*/, 6];
                case 6: return [2 /*return*/];
            }
        });
    });
}
exports.runExamples = runExamples;
// Run examples if this file is executed directly
if (require.main === module) {
    runExamples().then(function () {
        console.log('\n=== EXAMPLES COMPLETE ===');
    }).catch(console.error);
}
