// mkcentral-api/dist/lib/legacy/Players.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Players = void 0;
var node_url_1 = require("node:url");
var Player_1 = require("./Player");
var HTTP = __importStar(require("../common/APIRequest"));
var Players = /** @class */ (function (_super) {
    __extends(Players, _super);
    /**
     *
     * @param {string} rawData Data returned from the API as string.
     *
     * Providing no parameter creates an empty list of teams with default options
     */
    function Players(rawData) {
        var _this = _super.call(this) || this;
        if (!rawData) {
            _this.Count = 0;
        }
        else {
            var parsedData = JSON.parse(rawData);
            _this.Count = parsedData.count;
            for (var _i = 0, _a = parsedData.data; _i < _a.length; _i++) {
                var player = _a[_i];
                _this.push(new Player_1.Player(player));
            }
        }
        _this.Options = new Players.Options();
        return _this;
    }
    /**
     * Sends a request to the API asynchronously with parameters taken of local Options and updates the local object
     */
    Players.prototype.Get = function () {
        return __awaiter(this, void 0, void 0, function () {
            var uri, result, _a, _i, result_1, item;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        uri = new node_url_1.URL("https://www.mariokartcentral.com/mkc/api/registry/teams/category/".concat(this.Options.Category.toString()));
                        uri.searchParams.append('order', this.Options.Order.toString());
                        uri.searchParams.append('country', this.Options.Country.toString());
                        uri.searchParams.append('search', this.Options.Search.toString());
                        while (this.length) {
                            this.pop();
                        }
                        _a = Players.bind;
                        return [4 /*yield*/, HTTP.fetchHttpsContent(uri)];
                    case 1:
                        result = new (_a.apply(Players, [void 0, _b.sent()]))();
                        this.Count = result.Count;
                        for (_i = 0, result_1 = result; _i < result_1.length; _i++) {
                            item = result_1[_i];
                            this.push(item);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Sends a request to the API asynchronously with the given parameters and returns the result as a new Teams object
     * @param {Players.Options | Category | string | any} options
     *
     * Examples:
     * ```js
     * let result = await Players.Get();
     * let result = await Players.Get('200cc');
     * let result = await Players.Get(Players.Options.Category.SHADOW);
     * let result = await Players.Get(new Players.Options());
     * let result = await Players.Get(new Players.Options({
     *     Order: Order.REGISTRATION_TEAM_ASC,
     *     Country: Country.Unknown
     * }));
     * ```
     */
    Players.Get = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var o, uri, result, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        o = new Players.Options();
                        // if (options && options instanceof Players.Options.Category) {
                        //   o.Category = options;
                        // } else
                        if (options && options instanceof Players.Options) {
                            o = options;
                        }
                        else if (options && typeof options == 'string') {
                            o.Search = options;
                        }
                        uri = new node_url_1.URL("https://www.mariokartcentral.com/mkc/api/registry/players/category/".concat(o.Category.toString()));
                        uri.searchParams.append('order', o.Order.toString());
                        uri.searchParams.append('country', o.Country.toString());
                        uri.searchParams.append('search', o.Search.toString());
                        _a = Players.bind;
                        return [4 /*yield*/, HTTP.fetchHttpsContent(uri)];
                    case 1:
                        result = new (_a.apply(Players, [void 0, _b.sent()]))();
                        result.Options = o;
                        return [2 /*return*/, result];
                }
            });
        });
    };
    return Players;
}(Array));
exports.Players = Players;
(function (Players) {
    var Options = /** @class */ (function () {
        function Options(options) {
            this.Category = options ? (options.Category ? options.Category : Options.Category.ALL) : Options.Category.ALL;
            this.Country = options ? (options.Country ? options.Country : Options.Country.ALL) : Options.Country.ALL;
            this.Order = options ? (options.Order ? options.Order : Options.Order.NAME_ASC) : Options.Order.NAME_ASC;
            this.Search = options ? (options.Search ? options.Search : '') : '';
        }
        return Options;
    }());
    Players.Options = Options;
    (function (Options) {
        /**
         * Categories for players appended to .../registry/players/category/
         *
         * Default: .../registry/players/category/150cc
         */
        var Category;
        (function (Category) {
            /**
             * All Players
             */
            Category["ALL"] = "all";
            /**
             * All Players in MK8DX 150cc Team
             */
            Category["150CC"] = "150cc";
            /**
             * All Players in MK8DX 200cc Team
             */
            Category["200CC"] = "200cc";
            /**
             * All Players in MKTour VS Race Team
             */
            Category["MKTOUR"] = "mktour_vs";
            /**
             * All Players in no Team
             */
            Category["SHADOW"] = "shadow";
        })(Category = Options.Category || (Options.Category = {}));
        /**
         * Countries to filter players as parameter
         *
         * Default: country=ZZ
         */
        var Country;
        (function (Country) {
            /**
             * Every available language
             */
            Country["ALL"] = "ZZ";
            Country["Afghanistan"] = "AF";
            Country["Albania"] = "AL";
            Country["Algeria"] = "DZ";
            Country["AmericanSamoa"] = "AS";
            Country["Andorra"] = "AD";
            Country["Antarctica"] = "AQ";
            Country["Argentina"] = "AR";
            Country["Armenia"] = "AM";
            Country["Aruba"] = "AW";
            Country["Australia"] = "AU";
            Country["Austria"] = "AT";
            Country["Azerbaijan"] = "AZ";
            Country["Bahamas"] = "BS";
            Country["Bahrain"] = "BH";
            Country["Bangladesh"] = "BD";
            Country["Barbados"] = "BB";
            Country["Belarus"] = "BY";
            Country["Belgium"] = "BE";
            Country["Bolivia"] = "BO";
            Country["BonaireSintEustatiusAndSaba"] = "BQ";
            Country["BosniaAndHerzegovina"] = "BA";
            Country["Brazil"] = "BR";
            Country["BritishIndianOceanTerritory"] = "IO";
            Country["Canada"] = "CA";
            Country["Chile"] = "CL";
            Country["China"] = "CN";
            Country["Colombia"] = "CO";
            Country["Congo"] = "CG";
            Country["CostaRica"] = "CR";
            Country["Croatia"] = "HR";
            Country["Cuba"] = "CU";
            Country["Cyprus"] = "CY";
            Country["CzechRepublic"] = "CZ";
            Country["Denmark"] = "DK";
            Country["DominicanRepublic"] = "DO";
            Country["Ecuador"] = "EC";
            Country["Egypt"] = "EG";
            Country["ElSalvador"] = "SV";
            Country["Estonia"] = "EE";
            Country["Finland"] = "FI";
            Country["France"] = "FR";
            Country["FrenchPolynesia"] = "PF";
            Country["Georgia"] = "GE";
            Country["Germany"] = "DE";
            Country["Greece"] = "GR";
            Country["Guam"] = "GU";
            Country["Guatemala"] = "GT";
            Country["Honduras"] = "HN";
            Country["HongKong"] = "HK";
            Country["Hungary"] = "HU";
            Country["Iceland"] = "IS";
            Country["India"] = "IN";
            Country["Indonesia"] = "ID";
            Country["Ireland"] = "IE";
            Country["Israel"] = "IL";
            Country["Italy"] = "IT";
            Country["Jamaica"] = "JM";
            Country["Japan"] = "JP";
            Country["Jersey"] = "JE";
            Country["Jordan"] = "JO";
            Country["Kenya"] = "KE";
            Country["Latvia"] = "LV";
            Country["Lebanon"] = "LB";
            Country["Luxembourg"] = "LU";
            Country["Macao"] = "MO";
            Country["Madagascar"] = "MG";
            Country["Malaysia"] = "MY";
            Country["Malta"] = "MT";
            Country["Mexico"] = "MX";
            Country["Morocco"] = "MA";
            Country["Netherlands"] = "NL";
            Country["NewCaledonia"] = "NC";
            Country["NewZealand"] = "NZ";
            Country["Nicaragua"] = "NI";
            Country["Niger"] = "NE";
            Country["Norway"] = "NO";
            Country["Panama"] = "PA";
            Country["Paraguay"] = "PY";
            Country["Peru"] = "PE";
            Country["Philippines"] = "PH";
            Country["Poland"] = "PL";
            Country["Portugal"] = "PT";
            Country["PuertoRico"] = "PR";
            Country["Romania"] = "RO";
            Country["Russia"] = "RU";
            Country["Reunion"] = "RE";
            Country["SaudiArabia"] = "SA";
            Country["SierraLeone"] = "LS";
            Country["Singapore"] = "SG";
            Country["Slovakia"] = "SK";
            Country["Slovenia"] = "SI";
            Country["SouthAfrica"] = "ZA";
            Country["SouthGeorgiaAndTheSouthSandwichIslands"] = "GS";
            Country["SouthKorea"] = "KR";
            Country["Spain"] = "ES";
            Country["SriLanka"] = "LK";
            Country["Sudan"] = "SD";
            Country["Sweden"] = "SE";
            Country["Switzerland"] = "CH";
            Country["Taiwan"] = "TW";
            Country["Thailand"] = "TH";
            Country["TrinidadAndTobago"] = "TT";
            Country["Tunisia"] = "TN";
            Country["Turkey"] = "TR";
            Country["USVirginIslands"] = "VI";
            Country["Ukraine"] = "UA";
            Country["UnitedArabEmirates"] = "AE";
            Country["UnitedKingdom"] = "GB";
            Country["UnitedStates"] = "US";
            /**
             * Every unknown language
             */
            Country["Unknown"] = "XX";
            Country["Uruguay"] = "UY";
            Country["Venezuela"] = "VE";
            Country["Zimbabwe"] = "ZW";
        })(Country = Options.Country || (Options.Country = {}));
        /**
         * Sort parameter for player queries
         *
         * Default: order=NA
         */
        var Order;
        (function (Order) {
            /**
             * DEFAULT
             *
             * Order by name ascending.
             *
             * Valid in every category.
             */
            Order["NAME_ASC"] = "NA";
            /**
             * Order by name descending.
             *
             * Valid in every category.
             */
            Order["NAME_DESC"] = "ND";
            /**
             * Order by friendcode ascending.
             *
             * Uses the `custom_field`, which is the FC of the corresponding game to the category.
             */
            Order["FC_ASC"] = "FA";
            /**
             * Order by friendcode descending.
             *
             * Uses the `custom_field`, which is the FC of the corresponding game to the category.
             */
            Order["FC_DESC"] = "FD";
            /**
             * Order by switch friendcode ascending.
             *
             * Valid categories: all, shadow
             */
            // FC_SWITCH_ASC = 'SFA',
            /**
             * Order by switch friendcode descending.
             *
             * Valid categories: all, shadow
             */
            // FC_SWITCH_DESC = 'SFD',
            /**
             * Order by mktour friendcode ascending.
             *
             * Valid categories: all, shadow
             */
            // FC_TOUR_ASC = 'TFA',
            /**
             * Order by mktour friendcode descending.
             *
             * Valid categories: all, shadow
             */
            // FC_TOUR_DESC = 'TFD',
            /**
             * Order by 3DS friendcode ascending.
             *
             * Valid categories: all, shadow
             */
            // FC_3DS_ASC = '3FA',
            /**
             * Order by 3DS friendcode descending.
             *
             * Valid categories: all, shadow
             */
            // FC_3DS_DESC = '3FD',
            /**
             * Order by NNID ascending.
             *
             * Valid categories: all, shadow
             */
            // FC_NNID_ASC = 'UFA',
            /**
             * Order by NNID descending.
             *
             * Valid categories: all, shadow
             */
            // FC_NNID_DESC = 'UFD',
            /**
             * Order by latest team registration.
             *
             * Valid categories: 150cc, 200cc, mktour
             */
            Order["REGISTRATION_TEAM_ASC"] = "TA";
            /**
             * Order by latest team registration.
             *
             * Valid categories: 150cc, 200cc, mktour
             */
            Order["REGISTRATION_TEAM_DESC"] = "TD";
            /**
             * Order by player registration date ascending.
             *
             * Valid in every category.
             */
            Order["REGISTRATION_PLAYER_ASC"] = "RA";
            /**
             * Order by player registration date descending.
             *
             * Valid in every category.
             */
            Order["REGISTRATION_PLAYER_DESC"] = "RD";
        })(Order = Options.Order || (Options.Order = {}));
    })(Options = Players.Options || (Players.Options = {}));
})(Players = exports.Players || (exports.Players = {}));
exports.Players = Players;
