// mkcentral-api/dist/lib/legacy/Player.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Player = void 0;
var HTTP = __importStar(require("../common/APIRequest"));
var node_url_1 = require("node:url");
var Team_1 = require("./Team");
var Player = /** @class */ (function () {
    function Player(playerData) {
        if (typeof playerData == 'string') {
            var data = JSON.parse(playerData);
            this.Id = data.id;
            this.UserId = data.user_id;
            this.RegisteredDate = new Date(data.registered_at.date);
            this.RegisteredDateHuman = data.registered_at_human;
            this.TeamRegisteredDate = null;
            this.TeamRegisteredDateHuman = null;
            this.Name = data.display_name;
            this.Status = data.player_status;
            this.IsBanned = data.is_banned;
            this.BanReason = data.ban_reason;
            this.IsHidden = data.is_hidden == 0 ? false : true;
            this.CountryCode = data.country_code;
            this.CountryName = data.country_name;
            this.Region = data.region;
            this.City = data.city;
            this.DiscordPrivacy = data.discord_privacy;
            this.DiscordTag = data.discord_tag;
            this.SwitchFC = data.switch_fc;
            this.NNID = data.nnid;
            this.FC3DS = data.fc_3ds;
            this.MKTourFC = data.mktour_fc;
            this.ProfilePicture = data.profile_picture;
            this.ProfilePictureBorderColor = data.profile_picture_border_color;
            this.ProfileMessage = data.profile_message;
            this.IsSupporter = data.is_supporter;
            this.IsAdministrator = data.is_administrator;
            this.IsModerator = data.is_moderator;
            this.IsGlobalEventAdmin = data.is_global_event_admin;
            this.IsGlobalEventMod = data.is_global_event_mod;
            this.IsEventAdmin = data.is_event_admin;
            this.IsEventMod = data.is_event_mod;
            this.CurrentTeams = [];
            for (var _i = 0, _a = data.current_teams; _i < _a.length; _i++) {
                var team = _a[_i];
                this.CurrentTeams.push(new Player.CurrentTeam(team));
            }
        }
        else {
            this.Id = playerData.player_id;
            this.UserId = playerData.user_id;
            this.RegisteredDate = new Date(playerData.registered_at);
            this.RegisteredDateHuman = playerData.registered_at_human;
            this.TeamRegisteredDate = !playerData.team_registered_at ? null : new Date(playerData.team_registered_at);
            this.TeamRegisteredDateHuman = !playerData.team_registered_at_human ? null : playerData.team_registered_at_human;
            this.Name = playerData.display_name;
            this.Status = playerData.player_status;
            this.CountryCode = playerData.country_code;
            this.CountryName = playerData.country_name;
            this.SwitchFC = playerData.switch_fc;
            this.NNID = playerData.nnid;
            this.FC3DS = playerData.fc_3ds;
            this.MKTourFC = playerData.mktour_fc;
        }
    }
    Player.prototype.Load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var player;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, Player.Get(this.Id)];
                    case 1:
                        player = _a.sent();
                        this.Id = player.Id;
                        this.UserId = player.UserId;
                        this.RegisteredDate = player.RegisteredDate;
                        this.RegisteredDateHuman = player.RegisteredDateHuman;
                        this.TeamRegisteredDate = null;
                        this.TeamRegisteredDateHuman = null;
                        this.Name = player.Name;
                        this.Status = player.Status;
                        this.IsBanned = player.IsBanned;
                        this.BanReason = player.BanReason;
                        this.IsHidden = player.IsHidden;
                        this.CountryCode = player.CountryCode;
                        this.CountryName = player.CountryName;
                        this.Region = player.Region;
                        this.City = player.City;
                        this.DiscordPrivacy = player.DiscordPrivacy;
                        this.DiscordTag = player.DiscordTag;
                        this.SwitchFC = player.SwitchFC;
                        this.NNID = player.NNID;
                        this.FC3DS = player.FC3DS;
                        this.MKTourFC = player.MKTourFC;
                        this.ProfilePicture = player.ProfilePicture;
                        this.ProfilePictureBorderColor = player.ProfilePictureBorderColor;
                        this.ProfileMessage = player.ProfileMessage;
                        this.IsSupporter = player.IsSupporter;
                        this.IsAdministrator = player.IsAdministrator;
                        this.IsModerator = player.IsModerator;
                        this.IsGlobalEventAdmin = player.IsGlobalEventAdmin;
                        this.IsGlobalEventMod = player.IsGlobalEventMod;
                        this.IsEventAdmin = player.IsEventAdmin;
                        this.IsEventMod = player.IsEventMod;
                        this.CurrentTeams = player.CurrentTeams;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Sends a request to the API asynchronously with the given parameters and returns the result as a new Team object
     * @param {number} playerId
     *
     * Examples:
     * ```js
     * let result = await Team.Get(10);
     * ```
     */
    Player.Get = function (playerId) {
        return __awaiter(this, void 0, void 0, function () {
            var uri, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        uri = new node_url_1.URL("https://www.mariokartcentral.com/mkc/api/registry/players/".concat(playerId));
                        _a = Player.bind;
                        return [4 /*yield*/, HTTP.fetchHttpsContent(uri)];
                    case 1: return [2 /*return*/, new (_a.apply(Player, [void 0, _b.sent()]))()];
                }
            });
        });
    };
    return Player;
}());
exports.Player = Player;
(function (Player) {
    var CurrentTeam = /** @class */ (function () {
        function CurrentTeam(teamData) {
            this.Mode = teamData.mode;
            this.ModeTitle = teamData.mode_title;
            this.ModeKey = teamData.mode_key;
            this.TeamId = teamData.team_id;
            this.TeamName = teamData.team_name;
            this.TeamTag = teamData.team_tag;
            this.TeamStatus = teamData.team_status;
        }
        CurrentTeam.prototype.GetTeam = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, Team_1.Team.Get(this.TeamId)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        };
        return CurrentTeam;
    }());
    Player.CurrentTeam = CurrentTeam;
})(Player = exports.Player || (exports.Player = {}));
exports.Player = Player;
