// mkcentral-api/dist/lib/legacy/Teams.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Teams = void 0;
var node_url_1 = require("node:url");
var Team_1 = require("./Team");
var HTTP = __importStar(require("../common/APIRequest"));
var Teams = /** @class */ (function (_super) {
    __extends(Teams, _super);
    /**
     *
     * @param {string} rawData Data returned from the API as string.
     *
     * Providing no parameter creates an empty list of teams with default options
     */
    function Teams(rawData) {
        var _this = _super.call(this) || this;
        if (!rawData) {
            _this.Count = 0;
        }
        else {
            var parsedData = JSON.parse(rawData);
            _this.Count = parsedData.count;
            for (var _i = 0, _a = parsedData.data; _i < _a.length; _i++) {
                var team = _a[_i];
                _this.push(new Team_1.Team(team));
            }
        }
        _this.Options = new Teams.Options();
        return _this;
    }
    /**
     * Sends a request to the API asynchronously with parameters taken of local Options and updates the local object
     */
    Teams.prototype.Get = function () {
        return __awaiter(this, void 0, void 0, function () {
            var uri, result, _a, _i, result_1, item;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        uri = new node_url_1.URL("https://www.mariokartcentral.com/mkc/api/registry/teams/category/".concat(this.Options.Category.toString()));
                        uri.searchParams.append('order', this.Options.Order.toString());
                        uri.searchParams.append('language', this.Options.Language.toString());
                        uri.searchParams.append('search', this.Options.Search.toString());
                        while (this.length) {
                            this.pop();
                        }
                        _a = Teams.bind;
                        return [4 /*yield*/, HTTP.fetchHttpsContent(uri)];
                    case 1:
                        result = new (_a.apply(Teams, [void 0, _b.sent()]))();
                        this.Count = result.Count;
                        for (_i = 0, result_1 = result; _i < result_1.length; _i++) {
                            item = result_1[_i];
                            this.push(item);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Sends a request to the API asynchronously with the given parameters and returns the result as a new Teams object
     * @param {TeamFilter | Category | string | any} options
     *
     * Examples:
     * ```js
     * let result = await Teams.Get();
     * let result = await Teams.Get('200cc');
     * let result = await Teams.Get(Teams.Options.Category.HISTORICAL);
     * let result = await Teams.Get(new Teams.Options());
     * let result = await Teams.Get(new Teams.Options({
     *     Order: Order.REGISTRATION_ASC,
     *     Language: Language.OTHER
     * }));
     * ```
     */
    Teams.Get = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var o, uri, result, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        o = new Teams.Options();
                        if (options && options instanceof Teams.Options) {
                            o = options;
                        }
                        else if (options && typeof options == 'string') {
                            o.Search = options;
                        }
                        uri = new node_url_1.URL("https://www.mariokartcentral.com/mkc/api/registry/teams/category/".concat(o.Category.toString()));
                        uri.searchParams.append('order', o.Order.toString());
                        uri.searchParams.append('language', o.Language.toString());
                        uri.searchParams.append('search', o.Search.toString());
                        _a = Teams.bind;
                        return [4 /*yield*/, HTTP.fetchHttpsContent(uri)];
                    case 1:
                        result = new (_a.apply(Teams, [void 0, _b.sent()]))();
                        result.Options = o;
                        return [2 /*return*/, result];
                }
            });
        });
    };
    return Teams;
}(Array));
exports.Teams = Teams;
(function (Teams) {
    var Options = /** @class */ (function () {
        function Options(options) {
            this.Category = options
                ? options.Category
                    ? options.Category
                    : Options.Category.ACTIVE
                : Options.Category.ACTIVE;
            this.Language = options ? (options.Language ? options.Language : Options.Language.ALL) : Options.Language.ALL;
            this.Order = options ? (options.Order ? options.Order : Options.Order.NAME_ASC) : Options.Order.NAME_ASC;
            this.Search = options ? (options.Search ? options.Search : '') : '';
        }
        return Options;
    }());
    Teams.Options = Options;
    (function (Options) {
        /**
         * Categories for teams appended to .../registry/teams/category/
         *
         * Default: .../registry/teams/category/active
         */
        var Category;
        (function (Category) {
            /**
             * All Active Teams
             */
            Category["ALL"] = "active";
            /**
             * All Active Teams
             */
            Category["ACTIVE"] = "active";
            /**
             * All Active 150cc Teams
             */
            Category["150CC"] = "150cc";
            /**
             * All Active 200cc Teams
             */
            Category["200CC"] = "200cc";
            /**
             * All Active MK Tour Teams
             */
            Category["MKTOUR"] = "mktour_vs";
            /**
             * All Inactive/Old Teams
             */
            Category["HISTORICAL"] = "historical";
        })(Category = Options.Category || (Options.Category = {}));
        /**
         * Languages to filter teams as parameter
         *
         * Default: language=all
         */
        var Language;
        (function (Language) {
            /**
             * Every available language
             */
            Language["ALL"] = "all";
            /**
             * English teams only
             */
            Language["ENGLISH"] = "English";
            /**
             * French teams only
             */
            Language["FRENCH"] = "French";
            /**
             * German teams only
             */
            Language["GERMAN"] = "German";
            /**
             * Portuguese teams only
             */
            Language["PORTUGUESE"] = "Portuguese";
            /**
             * Spanish teams only
             */
            Language["SPANISH"] = "Spanish";
            /**
             * Teams without default language
             */
            Language["OTHER"] = "other";
        })(Language = Options.Language || (Options.Language = {}));
        /**
         * Sort parameter for team queries
         *
         * Default: order=NA
         */
        var Order;
        (function (Order) {
            /**
             * DEFAULT
             *
             * Order by name ascending.
             *
             * Valid in every category.
             */
            Order["NAME_ASC"] = "NA";
            /**
             * Order by name descending.
             *
             * Valid in every category.
             */
            Order["NAME_DESC"] = "ND";
            /**
             * Order by latest team registration.
             *
             * Valid categories: 150cc, 200cc, mktour
             */
            Order["REGISTRATION_ASC"] = "RA";
            /**
             * Order by latest team registration.
             *
             * Valid categories: 150cc, 200cc, mktour
             */
            Order["REGISTRATION_DESC"] = "RD";
            /**
             * Order by player registration date ascending.
             *
             * Valid in every category.
             */
            Order["STATUS_ASC"] = "SA";
            /**
             * Order by player registration date descending.
             *
             * Valid in every category.
             */
            Order["STATUS_DESC"] = "SD";
        })(Order = Options.Order || (Options.Order = {}));
    })(Options = Teams.Options || (Teams.Options = {}));
})(Teams = exports.Teams || (exports.Teams = {}));
exports.Teams = Teams;
