// mkcentral-api/dist/lib/legacy/Team.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Team = void 0;
var HTTP = __importStar(require("../common/APIRequest"));
var node_url_1 = require("node:url");
var Player_1 = require("./Player");
var Team = /** @class */ (function () {
    function Team(teamData) {
        if (typeof teamData == 'string') {
            var data = JSON.parse(teamData);
            this.Id = data.id;
            this.Category = data.team_category;
            this.Description = data.team_description;
            this.MainLanguage = data.main_language;
            this.Name = data.team_name;
            this.Tag = data.team_tag;
            this.Color = data.team_color;
            this.Status = data.team_status;
            this.RecruitmentStatus = data.recruitment_status;
            this.IsHistorical = data.is_historical == 1 ? true : false;
            this.IsShadow = false;
            this.PlayerCount = 0; // build some logic to get all members of all rosters without duplicates
            this.FoundingDate = new Date(data.founding_date.date);
            this.FoundingDateHuman = data.founding_date_human;
            this.Logo = new node_url_1.URL("https://www.mariokartcentral.com/mkc/storage/".concat(data.team_logo));
            this.PrimaryTeamId = data.primary_team_id;
            this.PrimaryTeamName = data.primary_team_name;
            this.Rosters = new Team.Rosters(data.rosters);
            this.SecondaryTeams = [];
            for (var _i = 0, _a = data.secondary_teams; _i < _a.length; _i++) {
                var team = _a[_i];
                this.SecondaryTeams.push({
                    Id: team.id,
                    Name: team.name,
                });
            }
        }
        else {
            this.Id = teamData.team_id;
            this.Name = teamData.team_name;
            this.Tag = teamData.team_tag;
            this.Color = teamData.team_color;
            this.Status = teamData.team_status;
            this.RecruitmentStatus = teamData.recruitment_status;
            this.IsHistorical = false;
            this.IsShadow = teamData.is_shadow == 1 ? true : false;
            this.PlayerCount = teamData.player_count;
            this.FoundingDate = new Date(teamData.founding_date);
            this.FoundingDateHuman = teamData.founding_date_human;
            this.SecondaryTeams = [];
        }
    }
    Team.prototype.Load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var team;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, Team.Get(this.Id)];
                    case 1:
                        team = _a.sent();
                        this.Id = team.Id;
                        this.Category = team.Category;
                        this.Description = team.Description;
                        this.MainLanguage = team.MainLanguage;
                        this.Name = team.Name;
                        this.Tag = team.Tag;
                        this.Color = team.Color;
                        this.Status = team.Status;
                        this.RecruitmentStatus = team.RecruitmentStatus;
                        this.IsHistorical = team.IsHistorical;
                        this.IsShadow = false;
                        this.PlayerCount = 0; // build some logic to get all members of all rosters without duplicates
                        this.FoundingDate = team.FoundingDate;
                        this.FoundingDateHuman = team.FoundingDateHuman;
                        this.Logo = team.Logo;
                        this.PrimaryTeamId = team.PrimaryTeamId;
                        this.PrimaryTeamName = team.PrimaryTeamName;
                        this.Rosters = team.Rosters;
                        this.SecondaryTeams = team.SecondaryTeams;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Sends a request to the API asynchronously with the given parameters and returns the result as a new Team object
     * @param {number} teamId
     *
     * Examples:
     * ```js
     * let result = await Team.Get(1064);
     * ```
     */
    Team.Get = function (teamId) {
        return __awaiter(this, void 0, void 0, function () {
            var uri, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        uri = new node_url_1.URL("https://www.mariokartcentral.com/mkc/api/registry/teams/".concat(teamId));
                        _a = Team.bind;
                        return [4 /*yield*/, HTTP.fetchHttpsContent(uri)];
                    case 1: return [2 /*return*/, new (_a.apply(Team, [void 0, _b.sent()]))()];
                }
            });
        });
    };
    return Team;
}());
exports.Team = Team;
(function (Team) {
    var Rosters = /** @class */ (function () {
        function Rosters(rawData) {
            this['150cc'] = rawData['150cc'] ? new Roster(rawData['150cc']) : null;
            this['200cc'] = rawData['200cc'] ? new Roster(rawData['200cc']) : null;
            this['mktour_vs'] = rawData['mktour_vs'] ? new Roster(rawData['mktour_vs']) : null;
        }
        return Rosters;
    }());
    Team.Rosters = Rosters;
    var Roster = /** @class */ (function () {
        function Roster(rawData) {
            this.ModeKey = rawData.mode_key;
            this.ModeTitle = rawData.mode_title;
            this.RosterName = rawData.roster_name;
            this.IsUserMember = rawData.is_user_member;
            this.IsActive = rawData.active == 1 ? true : false;
            this.Members = [];
            for (var _i = 0, _a = rawData.members; _i < _a.length; _i++) {
                var member = _a[_i];
                this.Members.push(new Member(member));
            }
        }
        return Roster;
    }());
    Team.Roster = Roster;
    var Member = /** @class */ (function () {
        function Member(rawData) {
            this.Id = rawData.player_id;
            this.Name = rawData.display_name;
            this.CustomFieldName = rawData.custom_field_name;
            this.CustomFieldValue = rawData.custom_field;
            this.Status = rawData.player_status;
            this.RegisteredDate = new Date(rawData.registered_since);
            this.RegisteredDateHuman = rawData.registered_since_human;
            this.CountryCode = rawData.country_code;
            this.CountryName = rawData.country_name;
            this.IsTeamLeader = rawData.team_leader == 1 ? true : false;
        }
        Member.prototype.GetPlayer = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, Player_1.Player.Get(this.Id)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        };
        return Member;
    }());
    Team.Member = Member;
})(Team = exports.Team || (exports.Team = {}));
exports.Team = Team;
